// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/domain/service/EntryDataValidator.js
import { jsonSchemaToZod, zodPathToJsonPointer, } from "./JsonSchemaToZod.js";
import { runtimeDiagnostic } from "../../kernel/diagnostic.js";
/**
 * Per-collection entry-data validator.
 *
 * Given a parsed `SchemaManifest` and an entry payload, runs zod
 * validation against the manifest's `spec.schema` and returns
 * Diagnostics. Compiled zod schemas are cached per `metadata.name` so
 * repeated `validate()` calls are cheap; the cache is content-
 * addressed-ish — calling `validate` with two different
 * SchemaManifests that share a `metadata.name` will reuse the first
 * compile, which is correct because manifest names are unique within a
 * deployment.
 *
 * Switched from Ajv to zod (POC issue #70) because Ajv's runtime
 * `compile()` calls `new Function()`, which CF Workers blocks under
 * V8's codegen-from-strings policy. zod composes builders into a
 * plain object tree and runs them through its interpreter at parse
 * time — no codegen, Workers-safe.
 *
 * The `x-clam-ref`, `x-mcp-hint`, `x-clam-bind` extension keywords
 * are tolerated (the converter ignores unknown keywords; they pass
 * through unmodified).
 *
 * Output shape is `Diagnostic[]` — empty array means "valid". The
 * dispatcher and admin-write paths surface the array directly; the
 * `phase: "runtime"` stamp is correct for entry-write validation,
 * which always happens at request-handling time.
 */
export class EntryDataValidator {
    compiled = new Map();
    /**
     * Validate `data` against `manifest.spec.schema`. Returns the empty
     * array on success; on failure, returns one or more Diagnostics
     * with `code: "INPUT_VALIDATION_FAILED"` and `path` set to the
     * RFC 6901 JSON Pointer of the offending field (`""` = root).
     */
    validate(manifest, data) {
        const compiled = this.compileFor(manifest);
        const result = compiled.safeParse(data);
        if (result.success)
            return [];
        return result.error.issues.map((issue) => runtimeDiagnostic({
            code: "INPUT_VALIDATION_FAILED",
            severity: "error",
            path: zodPathToJsonPointer(issue.path),
            message: issue.message,
        }));
    }
    /**
     * Test-only escape hatch: peek at whether a given Schema has been
     * compiled yet. The cache is otherwise an internal implementation
     * detail. Not part of the package's public API surface.
     *
     * @internal
     */
    hasCompiled(name) {
        return this.compiled.has(name);
    }
    compileFor(manifest) {
        const name = manifest.metadata.name;
        let compiled = this.compiled.get(name);
        if (!compiled) {
            compiled = jsonSchemaToZod(manifest.spec.schema);
            this.compiled.set(name, compiled);
        }
        return compiled;
    }
}
/** Backwards-compatible alias for the v0.1.0 import name. Will be
 *  dropped once consumers migrate to `EntryDataValidator`. */
export { EntryDataValidator as SchemaValidator };
