// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/domain/service/LifecycleStateMachine.js
const DEFAULT_LIFECYCLE = "simple";
export function resolveLifecycle(schema) {
    return schema?.spec.lifecycle ?? DEFAULT_LIFECYCLE;
}
/**
 * Whether `requestPublish` on an entry of this Schema should write an
 * approval row and flip the entry to `'review'` (editorial) versus
 * publishing immediately and skipping the approval table (simple).
 */
export function publishRequiresApproval(schema) {
    return resolveLifecycle(schema) === "editorial";
}
/**
 * Allowed status transitions per lifecycle. Used by the MCP handler
 * and admin endpoints to gate operations. Unknown transitions return
 * `false` and the caller should reject with `CONFLICT`.
 *
 * Simple lifecycle:
 *   draft → published, draft → archived
 *   published → archived, published → draft (unpublish-as-edit)
 *   archived → draft
 *
 * Editorial lifecycle:
 *   draft → review, draft → archived
 *   review → approved, review → draft (rejected)
 *   approved → scheduled, approved → published
 *   scheduled → published, scheduled → draft
 *   published → archived, published → draft
 *   archived → draft
 */
export function canTransition(schema, from, to) {
    const allowed = transitionsFor(resolveLifecycle(schema));
    return allowed[from]?.has(to) ?? false;
}
const SIMPLE_TRANSITIONS = {
    draft: new Set(["published", "archived"]),
    review: new Set(),
    approved: new Set(),
    scheduled: new Set(),
    published: new Set(["archived", "draft"]),
    archived: new Set(["draft"]),
};
const EDITORIAL_TRANSITIONS = {
    draft: new Set(["review", "archived"]),
    review: new Set(["approved", "draft"]),
    approved: new Set(["scheduled", "published"]),
    scheduled: new Set(["published", "draft"]),
    published: new Set(["archived", "draft"]),
    archived: new Set(["draft"]),
};
function transitionsFor(mode) {
    return mode === "editorial" ? EDITORIAL_TRANSITIONS : SIMPLE_TRANSITIONS;
}
/**
 * The **navigable** subset of statuses an admin user filters by in
 * the per-Schema sub-nav (Drafts / Review / Published / Archived for
 * editorial; same minus Review for simple). This is intentionally a
 * subset of the full state set: `approved` and `scheduled` are
 * intermediate editorial states — entries pass through them but
 * authors don't navigate to them as buckets.
 *
 * Lives in spec for now because the consumer's admin SPA bundles
 * this. v0.1.x may move to `clam-cms-admin-ui` once the SPA is
 * extracted from the spec dependency graph.
 */
export function getLifecycleStatuses(mode) {
    return mode === "editorial"
        ? ["draft", "review", "published", "archived"]
        : ["draft", "published", "archived"];
}
