// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/domain/service/ManifestParser.js
import { parseAllDocuments } from "yaml";
import { validateDiagnostic, } from "../../kernel/diagnostic.js";
import { API_VERSION, BUILTIN_OPS, LIFECYCLE_HOOKS, VIEW_PARAMS_RESERVED, isParamRef, } from "../model/ManifestGrammar.js";
/**
 * Day-1 envelope-and-shape parser. Loop 1 (`clam-cms validate`) does
 * the cross-manifest checks (Trigger.target.procedure exists, View.from
 * is a Schema, etc.) — see ADR-0007 / `docs/authoring-contract.md`.
 *
 * Diagnostics emitted here are intentionally narrow: bad envelope,
 * structurally malformed spec, use of a DRAFT or v0.1.x-not-yet-shipped
 * key the v0.1.0 parser does not accept.
 *
 * Return shape is `{ manifests, diagnostics }`: parse-fatal docs are
 * skipped (manifest absent from `manifests`) and reported via a
 * `severity: "error"` diagnostic. Per ADR-0008 the caller (the CLI / boot
 * validator / consumer) routes diagnostics; we don't throw.
 *
 * Multi-doc YAML support per ADR-0001 § "Authoring shape" — one feature
 * per file, atoms separated by `---`.
 */
/**
 * Throwable carrier used by the envelope-shape validators below
 * (`validateEnvelope`, kind-specific `validate*Spec`). Each throw
 * carries a JSON Pointer + diagnostic code; the top-level
 * `parseManifests` catches the throw and converts it to a Diagnostic
 * for the public `{ manifests, diagnostics }` return shape.
 */
export class ManifestParseError extends Error {
    docIndex;
    pointer;
    code;
    constructor(message, docIndex, 
    /** JSON Pointer into the manifest (e.g. `/spec/output`). */
    pointer, code = "INVALID_MANIFEST_ENVELOPE") {
        super(docIndex != null ? `[doc ${docIndex}] ${message}` : message);
        this.docIndex = docIndex;
        this.pointer = pointer;
        this.code = code;
        this.name = "ManifestParseError";
    }
}
const KNOWN_KINDS = new Set([
    "Schema",
    "View",
    "Procedure",
    "Trigger",
]);
const V01_TRIGGER_SOURCE_KINDS = new Set(["http", "lifecycle"]);
const DRAFT_TRIGGER_SOURCE_KINDS = new Set([
    "mcp",
    "cron",
    "queue",
]);
const V01_HTTP_METHODS = new Set([
    "POST",
    "PUT",
    "PATCH",
    "DELETE",
]);
const V01_HANDLER_KINDS = new Set(["ref", "builtin"]);
const V01_BUILTIN_OPS = new Set(BUILTIN_OPS);
const V01_LIFECYCLE_HOOKS = new Set(LIFECYCLE_HOOKS);
const V01_HOOK_ERROR_POLICIES = new Set(["abort", "continue"]);
const V01_LIFECYCLE_MODES = new Set(["simple", "editorial"]);
/**
 * Parse YAML text (single doc, multi-doc, or a list of either) into
 * typed manifests + diagnostics. Per ADR-0001 multi-doc YAML support,
 * `---` separators inside one string yield one manifest per doc.
 */
export function parseManifests(input) {
    const inputs = typeof input === "string" ? [input] : input;
    const manifests = [];
    const diagnostics = [];
    let globalDocIndex = 0;
    for (const yamlText of inputs) {
        const docCount = parseOneStream(yamlText, globalDocIndex, manifests, diagnostics);
        globalDocIndex += docCount;
    }
    return { manifests, diagnostics };
}
/**
 * Convenience wrapper around `parseManifests` for the common case
 * where any diagnostic is fatal (worker module-init, CLI tools).
 * Throws an `Error` with one diagnostic per line —
 * `[CODE] path: message` — when `result.diagnostics.length > 0`.
 *
 * Adapters and starters should prefer this over hand-rolling the
 * "format diagnostics → throw" three-liner so the error envelope
 * stays consistent across consumers (matters for AI authors reading
 * boot failures in `wrangler tail`).
 */
export function parseManifestsOrThrow(input, options) {
    const result = parseManifests(input);
    if (result.diagnostics.length > 0) {
        const summary = result.diagnostics
            .map((d) => `  - [${d.code}] ${d.path}: ${d.message}`)
            .join("\n");
        const ctx = options?.context ? ` in ${options.context}` : "";
        throw new Error(`Manifest parse failed${ctx}:\n${summary}`);
    }
    return result.manifests;
}
function parseOneStream(yamlText, baseDocIndex, manifests, diagnostics) {
    const docs = parseAllDocuments(yamlText, { merge: false });
    for (let i = 0; i < docs.length; i++) {
        const doc = docs[i];
        const docIndex = baseDocIndex + i;
        if (doc.errors.length > 0) {
            diagnostics.push(validateDiagnostic({
                code: "INVALID_MANIFEST_ENVELOPE",
                severity: "error",
                path: pointerFor(docIndex, "/"),
                message: `[doc ${docIndex}] YAML parse error: ${doc.errors.map((e) => e.message).join("; ")}`,
            }));
            continue;
        }
        const value = doc.toJS({ maxAliasCount: -1 });
        if (value == null)
            continue;
        try {
            manifests.push(validateEnvelope(value, docIndex));
        }
        catch (e) {
            if (e instanceof ManifestParseError) {
                diagnostics.push(validateDiagnostic({
                    code: e.code,
                    severity: "error",
                    path: pointerFor(docIndex, e.pointer ?? "/"),
                    message: e.message,
                }));
            }
            else {
                diagnostics.push(validateDiagnostic({
                    code: "INVALID_MANIFEST_ENVELOPE",
                    severity: "error",
                    path: pointerFor(docIndex, "/"),
                    message: e instanceof Error
                        ? `[doc ${docIndex}] ${e.message}`
                        : `[doc ${docIndex}] unknown parse error`,
                }));
            }
        }
    }
    return docs.length;
}
function pointerFor(docIndex, jsonPointer) {
    const ptr = jsonPointer.startsWith("/") ? jsonPointer : `/${jsonPointer}`;
    return `manifest:doc/${docIndex}#${ptr}`;
}
function validateEnvelope(raw, docIndex) {
    if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
        throw new ManifestParseError("manifest must be a YAML mapping", docIndex);
    }
    const m = raw;
    if (m["apiVersion"] !== API_VERSION) {
        throw new ManifestParseError(`apiVersion must be "${API_VERSION}"; got ${JSON.stringify(m["apiVersion"])}`, docIndex, "/apiVersion");
    }
    const kind = m["kind"];
    if (typeof kind !== "string" || !KNOWN_KINDS.has(kind)) {
        throw new ManifestParseError(`kind must be one of ${[...KNOWN_KINDS].join(", ")}; got ${JSON.stringify(kind)}`, docIndex, "/kind");
    }
    const meta = m["metadata"];
    if (typeof meta !== "object" || meta === null) {
        throw new ManifestParseError("metadata is required and must be a mapping", docIndex, "/metadata");
    }
    const name = meta["name"];
    if (typeof name !== "string" || name.length === 0) {
        throw new ManifestParseError("metadata.name is required (non-empty string)", docIndex, "/metadata/name");
    }
    const spec = m["spec"];
    if (typeof spec !== "object" || spec === null) {
        throw new ManifestParseError("spec is required and must be a mapping", docIndex, "/spec");
    }
    switch (kind) {
        case "Schema":
            return validateSchemaSpec(raw, docIndex);
        case "View":
            return validateViewSpec(raw, docIndex);
        case "Procedure":
            return validateProcedureSpec(raw, docIndex);
        case "Trigger":
            return validateTriggerSpec(raw, docIndex);
        default:
            throw new ManifestParseError(`unhandled kind ${kind}`, docIndex);
    }
}
function validateSchemaSpec(m, idx) {
    const s = m.spec;
    if (typeof s["schema"] !== "object" || s["schema"] === null) {
        throw new ManifestParseError("Schema.spec.schema is required", idx, "/spec/schema");
    }
    if (typeof s["title"] !== "string" || s["title"].length === 0) {
        throw new ManifestParseError("Schema.spec.title is required (non-empty string). It's the admin UI label — populate it in the user's primary language, not the bare metadata.name.", idx, "/spec/title");
    }
    if ("localized" in s && typeof s["localized"] !== "boolean") {
        throw new ManifestParseError(`Schema.spec.localized must be a boolean; got ${JSON.stringify(s["localized"])}`, idx, "/spec/localized");
    }
    if ("lifecycle" in s) {
        const lc = s["lifecycle"];
        if (typeof lc !== "string" || !V01_LIFECYCLE_MODES.has(lc)) {
            throw new ManifestParseError(`Schema.spec.lifecycle must be one of ${[...V01_LIFECYCLE_MODES].join(", ")}; got ${JSON.stringify(lc)}`, idx, "/spec/lifecycle");
        }
    }
    if ("translates" in s && s["translates"] != null) {
        const t = s["translates"];
        if (typeof t !== "object" || Array.isArray(t)) {
            throw new ManifestParseError("Schema.spec.translates must be an object { parent, on }", idx, "/spec/translates");
        }
        const tr = t;
        if (typeof tr["parent"] !== "string" || tr["parent"].length === 0) {
            throw new ManifestParseError("Schema.spec.translates.parent is required (non-empty Schema name)", idx, "/spec/translates/parent");
        }
        if (typeof tr["on"] !== "string" || tr["on"].length === 0) {
            throw new ManifestParseError("Schema.spec.translates.on is required (non-empty field name)", idx, "/spec/translates/on");
        }
        if (s["localized"] !== true) {
            throw new ManifestParseError("Schema.spec.translates requires Schema.spec.localized: true (a non-localized translation table is meaningless)", idx, "/spec/translates");
        }
    }
    if ("policies" in s) {
        throw new ManifestParseError("Schema.spec.policies is DRAFT (see ADR-0001 § \"What's DRAFT\" / Schema); not supported in v0.1", idx, "/spec/policies", "DRAFT_KEY_USED");
    }
    return m;
}
function validateViewSpec(m, idx) {
    const s = m.spec;
    if (typeof s["from"] !== "string" || s["from"].length === 0) {
        throw new ManifestParseError("View.spec.from is required (non-empty string)", idx, "/spec/from");
    }
    let paramSchema;
    if ("params" in s && s["params"] != null) {
        paramSchema = validateViewParams(s["params"], idx);
    }
    if ("filter" in s && s["filter"] != null) {
        validateFilterAst(s["filter"], idx, "View.spec.filter", "/spec/filter", paramSchema);
    }
    for (const draft of ["recursive", "gatedBy", "join", "policies"]) {
        if (draft in s) {
            throw new ManifestParseError(`View.spec.${draft} is DRAFT (see ADR-0001 § "What's DRAFT" / View); not supported in v0.1`, idx, `/spec/${draft}`, "DRAFT_KEY_USED");
        }
    }
    return m;
}
function validateViewParams(raw, idx) {
    if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
        throw new ManifestParseError("View.spec.params must be a JSON Schema object", idx, "/spec/params", "VIEW_PARAMS_INVALID_SHAPE");
    }
    const p = raw;
    if (p["type"] !== "object") {
        throw new ManifestParseError(`View.spec.params.type must be "object"; got ${JSON.stringify(p["type"])}`, idx, "/spec/params/type", "VIEW_PARAMS_INVALID_SHAPE");
    }
    const props = p["properties"];
    if (typeof props !== "object" || props === null || Array.isArray(props)) {
        throw new ManifestParseError("View.spec.params.properties is required (declare each accepted query-string param)", idx, "/spec/params/properties", "VIEW_PARAMS_INVALID_SHAPE");
    }
    const propNames = Object.keys(props);
    for (const reserved of VIEW_PARAMS_RESERVED) {
        if (propNames.includes(reserved)) {
            throw new ManifestParseError(`View.spec.params.properties.${reserved} is reserved (the runtime owns ${VIEW_PARAMS_RESERVED.join(", ")} for pagination); rename the param.`, idx, `/spec/params/properties/${reserved}`, "VIEW_PARAMS_RESERVED_NAME");
        }
    }
    return raw;
}
function validateFilterAst(node, idx, path, jsonPointer, paramSchema) {
    if (typeof node !== "object" || node === null || Array.isArray(node)) {
        throw new ManifestParseError(`${path} must be an object node (eq | and | or)`, idx, jsonPointer);
    }
    const n = node;
    const keys = Object.keys(n);
    if (keys.length !== 1) {
        throw new ManifestParseError(`${path} must have exactly one key (eq | and | or); got ${JSON.stringify(keys)}`, idx, jsonPointer);
    }
    const op = keys[0];
    if (op === "eq") {
        const eq = n["eq"];
        if (typeof eq !== "object" || eq === null) {
            throw new ManifestParseError(`${path}.eq must be an object`, idx, `${jsonPointer}/eq`);
        }
        const e = eq;
        if (typeof e["field"] !== "string" || e["field"].length === 0) {
            throw new ManifestParseError(`${path}.eq.field is required (non-empty string)`, idx, `${jsonPointer}/eq/field`);
        }
        if (!("value" in e)) {
            throw new ManifestParseError(`${path}.eq.value is required`, idx, `${jsonPointer}/eq/value`);
        }
        if (isParamRef(e["value"])) {
            validateParamRef(e["value"].$param, idx, `${jsonPointer}/eq/value/$param`, paramSchema);
        }
        return;
    }
    if (op === "and" || op === "or") {
        const arr = n[op];
        if (!Array.isArray(arr) || arr.length === 0) {
            throw new ManifestParseError(`${path}.${op} must be a non-empty array`, idx, `${jsonPointer}/${op}`);
        }
        for (let i = 0; i < arr.length; i++) {
            validateFilterAst(arr[i], idx, `${path}.${op}[${i}]`, `${jsonPointer}/${op}/${i}`, paramSchema);
        }
        return;
    }
    const draftOps = new Set(["contains", "not", "in", "like"]);
    if (draftOps.has(op)) {
        throw new ManifestParseError(`${path} operator '${op}' is DRAFT (see ADR-0001 § "What's DRAFT" / View); not supported in v0.1`, idx, jsonPointer, "DRAFT_KEY_USED");
    }
    throw new ManifestParseError(`${path} operator must be one of eq, and, or; got '${op}'`, idx, jsonPointer);
}
function validateParamRef(name, idx, pointer, paramSchema) {
    if (name.length === 0) {
        throw new ManifestParseError("filter param ref { $param: '' } is invalid; supply a non-empty param name.", idx, pointer);
    }
    if (!paramSchema) {
        throw new ManifestParseError(`filter references param '${name}' but View.spec.params is not declared. Add a params JSON Schema or use a literal value.`, idx, pointer, "VIEW_FILTER_PARAM_REF_UNKNOWN");
    }
    const props = (paramSchema.properties ?? {});
    if (!Object.prototype.hasOwnProperty.call(props, name)) {
        throw new ManifestParseError(`filter references unknown param '${name}'; declare it under View.spec.params.properties.`, idx, pointer, "VIEW_FILTER_PARAM_REF_UNKNOWN");
    }
    const required = paramSchema.required ?? [];
    if (!required.includes(name)) {
        throw new ManifestParseError(`filter references optional param '${name}'; v0.1.0 requires every param-ref'd name to appear in View.spec.params.required (optional-skip semantics are reserved for v0.1.x).`, idx, pointer, "VIEW_FILTER_PARAM_REF_NOT_REQUIRED");
    }
}
function validateProcedureSpec(m, idx) {
    const s = m.spec;
    if (typeof s["input"] !== "object" || s["input"] === null) {
        throw new ManifestParseError("Procedure.spec.input is required (JSON Schema)", idx, "/spec/input");
    }
    if (typeof s["output"] !== "object" || s["output"] === null) {
        throw new ManifestParseError("Procedure.spec.output is required (JSON Schema)", idx, "/spec/output");
    }
    const handler = s["handler"];
    if (!handler) {
        throw new ManifestParseError("Procedure.spec.handler is required", idx, "/spec/handler");
    }
    validateHandlerBinding(handler, idx);
    if ("requires" in s && s["requires"] != null) {
        validateRequires(s["requires"], idx);
    }
    for (const draft of ["errors", "retry", "idempotency"]) {
        if (draft in s) {
            throw new ManifestParseError(`Procedure.spec.${draft} is DRAFT (see ADR-0001 § "What's DRAFT" / Procedure); not supported in v0.1`, idx, `/spec/${draft}`, "DRAFT_KEY_USED");
        }
    }
    return m;
}
function validateHandlerBinding(h, idx) {
    const kind = h["kind"];
    if (typeof kind !== "string") {
        throw new ManifestParseError(`Procedure.spec.handler.kind is required (one of ${[...V01_HANDLER_KINDS].join(", ")})`, idx, "/spec/handler/kind");
    }
    if (!V01_HANDLER_KINDS.has(kind)) {
        throw new ManifestParseError(`Procedure.spec.handler.kind must be one of ${[...V01_HANDLER_KINDS].join(", ")}; got '${kind}'`, idx, "/spec/handler/kind");
    }
    if (kind === "ref") {
        if (typeof h["ref"] !== "string" || h["ref"].length === 0) {
            throw new ManifestParseError("Procedure.spec.handler.ref is required (non-empty registration key)", idx, "/spec/handler/ref");
        }
        return;
    }
    const op = h["op"];
    if (typeof op !== "string" || !V01_BUILTIN_OPS.has(op)) {
        throw new ManifestParseError(`Procedure.spec.handler.op must be one of ${[...V01_BUILTIN_OPS].join(", ")}; got ${JSON.stringify(op)}`, idx, "/spec/handler/op");
    }
    if (typeof h["schema"] !== "string" || h["schema"].length === 0) {
        throw new ManifestParseError("Procedure.spec.handler.schema is required (Schema metadata.name) when handler.kind is 'builtin'", idx, "/spec/handler/schema");
    }
    if ("ref" in h) {
        throw new ManifestParseError("Procedure.spec.handler.ref is invalid when handler.kind is 'builtin' (ref + builtin are mutually exclusive)", idx, "/spec/handler/ref");
    }
}
function validateRequires(req, idx) {
    if (typeof req !== "object" || req === null) {
        throw new ManifestParseError("Procedure.spec.requires must be an object", idx);
    }
    const r = req;
    for (const draft of ["window", "quota"]) {
        if (draft in r) {
            throw new ManifestParseError(`Procedure.spec.requires.${draft} is DRAFT (see ADR-0001 § "What's DRAFT" / Procedure); not supported in v0.1`, idx, undefined, "DRAFT_KEY_USED");
        }
    }
    if (!("auth" in r) || r["auth"] == null)
        return;
    const auth = r["auth"];
    if (typeof auth !== "object" || auth === null) {
        throw new ManifestParseError("Procedure.spec.requires.auth must be an object", idx);
    }
    const a = auth;
    if ("any" in a) {
        throw new ManifestParseError("Procedure.spec.requires.auth.any is DRAFT; v0.1 supports only `all`", idx, undefined, "DRAFT_KEY_USED");
    }
    if (!("all" in a)) {
        throw new ManifestParseError("Procedure.spec.requires.auth must declare `all` (v0.1)", idx);
    }
    const all = a["all"];
    if (!Array.isArray(all) || all.length === 0) {
        throw new ManifestParseError("Procedure.spec.requires.auth.all must be a non-empty array", idx);
    }
    for (let i = 0; i < all.length; i++) {
        validateAuthPredicate(all[i], idx, `Procedure.spec.requires.auth.all[${i}]`);
    }
}
function validateAuthPredicate(p, idx, path) {
    if (p === "ctx.user")
        return;
    if (typeof p === "object" && p !== null && !Array.isArray(p)) {
        const o = p;
        if ("ctx.staff" in o) {
            const roles = o["ctx.staff"];
            if (!Array.isArray(roles) || roles.length === 0 || roles.some((r) => typeof r !== "string")) {
                throw new ManifestParseError(`${path}: 'ctx.staff' value must be a non-empty array of role-name strings`, idx);
            }
            return;
        }
    }
    if (typeof p === "object" && p !== null) {
        const draftKeys = ["owns", "withinMinutes", "contains"];
        const used = draftKeys.find((k) => k in p);
        if (used) {
            throw new ManifestParseError(`${path}: predicate '${used}' is DRAFT (see ADR-0001 § "What's DRAFT" / Procedure); not supported in v0.1`, idx, undefined, "DRAFT_KEY_USED");
        }
    }
    throw new ManifestParseError(`${path} must be 'ctx.user' or { 'ctx.staff': [<role>, ...] }; got ${JSON.stringify(p)}`, idx);
}
function validateHttpSource(source, idx) {
    const method = source["method"];
    if (typeof method !== "string" || !V01_HTTP_METHODS.has(method)) {
        throw new ManifestParseError(`Trigger.spec.source.method must be one of ${[...V01_HTTP_METHODS].join(", ")} (v0.1); got ${JSON.stringify(method)}`, idx, "/spec/source/method");
    }
    const path = source["path"];
    if (typeof path !== "string" || path.length === 0 || !path.startsWith("/")) {
        throw new ManifestParseError("Trigger.spec.source.path is required (non-empty string starting with '/')", idx, "/spec/source/path");
    }
}
function validateLifecycleSource(source, idx) {
    if (typeof source["schema"] !== "string" || source["schema"].length === 0) {
        throw new ManifestParseError("Trigger.spec.source.schema is required (Schema metadata.name) when source.kind is 'lifecycle'", idx, "/spec/source/schema");
    }
    const on = source["on"];
    if (!Array.isArray(on) || on.length === 0) {
        throw new ManifestParseError(`Trigger.spec.source.on must be a non-empty array of hook names (one of ${[...V01_LIFECYCLE_HOOKS].join(", ")})`, idx, "/spec/source/on");
    }
    for (let i = 0; i < on.length; i++) {
        const hook = on[i];
        if (typeof hook !== "string" || !V01_LIFECYCLE_HOOKS.has(hook)) {
            throw new ManifestParseError(`Trigger.spec.source.on[${i}] must be one of ${[...V01_LIFECYCLE_HOOKS].join(", ")}; got ${JSON.stringify(hook)}`, idx, `/spec/source/on/${i}`);
        }
    }
    if ("errorPolicy" in source) {
        const ep = source["errorPolicy"];
        if (typeof ep !== "string" || !V01_HOOK_ERROR_POLICIES.has(ep)) {
            throw new ManifestParseError(`Trigger.spec.source.errorPolicy must be 'abort' or 'continue'; got ${JSON.stringify(ep)}`, idx, "/spec/source/errorPolicy");
        }
        if (ep === "abort" && on.every((h) => typeof h === "string" && h.startsWith("after_"))) {
            throw new ManifestParseError("Trigger.spec.source.errorPolicy: 'abort' is invalid on after_* hooks — after_* runs after the response is sent, so abort cannot reach the caller. Move the hook to before_*, or use 'continue'.", idx, "/spec/source/errorPolicy");
        }
    }
    if ("method" in source || "path" in source) {
        throw new ManifestParseError("Trigger.spec.source.{method,path} are invalid when source.kind is 'lifecycle' (those keys belong to source.kind: 'http')", idx, "/spec/source");
    }
}
function validateTriggerSpec(m, idx) {
    const s = m.spec;
    const source = s["source"];
    if (!source) {
        throw new ManifestParseError("Trigger.spec.source is required", idx, "/spec/source");
    }
    const sourceKind = source["kind"];
    if (typeof sourceKind !== "string") {
        throw new ManifestParseError(`Trigger.spec.source.kind is required (one of ${[...V01_TRIGGER_SOURCE_KINDS].join(", ")})`, idx, "/spec/source/kind");
    }
    if (DRAFT_TRIGGER_SOURCE_KINDS.has(sourceKind)) {
        throw new ManifestParseError(`Trigger.spec.source.kind '${sourceKind}' is DRAFT (see ADR-0001 § "What's DRAFT" / Trigger); not supported in v0.1`, idx, "/spec/source/kind", "DRAFT_KEY_USED");
    }
    if (!V01_TRIGGER_SOURCE_KINDS.has(sourceKind)) {
        throw new ManifestParseError(`Trigger.spec.source.kind must be one of ${[...V01_TRIGGER_SOURCE_KINDS].join(", ")}; got '${sourceKind}'`, idx, "/spec/source/kind");
    }
    if (sourceKind === "http")
        validateHttpSource(source, idx);
    else if (sourceKind === "lifecycle")
        validateLifecycleSource(source, idx);
    const target = s["target"];
    if (!target || typeof target["procedure"] !== "string") {
        throw new ManifestParseError("Trigger.spec.target.procedure is required (string)", idx, "/spec/target/procedure");
    }
    if ("project" in target) {
        throw new ManifestParseError("Trigger.spec.target.project is DRAFT (see ADR-0001 § \"What's DRAFT\" / Trigger); not supported in v0.1", idx, undefined, "DRAFT_KEY_USED");
    }
    if ("atomicity" in s) {
        throw new ManifestParseError("Trigger.spec.atomicity is DRAFT (see ADR-0001 § \"What's DRAFT\" / Trigger); not supported in v0.1", idx, undefined, "DRAFT_KEY_USED");
    }
    return m;
}
/**
 * Bucketize a flat `Manifest[]` (parser output) into typed per-kind
 * arrays. Lives next to the parser because every caller pairs it with
 * `parseManifests` — they're the parser-output pipeline.
 */
export function partitionManifests(manifests) {
    const schemas = [];
    const views = [];
    const procedures = [];
    const triggers = [];
    for (const m of manifests) {
        if (m.kind === "Schema")
            schemas.push(m);
        else if (m.kind === "View")
            views.push(m);
        else if (m.kind === "Procedure")
            procedures.push(m);
        else
            triggers.push(m);
    }
    return { schemas, views, procedures, triggers };
}
