// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/lifecycle/RunLifecycleHooksUseCase.js
import { DiagnosticError, runtimeDiagnostic, } from "@aotterclam/clam-cms-spec";
/**
 * `RunLifecycleHooksUseCase` — implements the `LifecycleHookRunner`
 * port. Resolves matching Triggers via `TriggerIndex`, dispatches each
 * to its target Procedure through `InvokeProcedureUseCase` (so input
 * validation, auth, output validation all reuse the shared pipeline),
 * and applies per-Trigger `errorPolicy`.
 *
 * Defaults from the parser-locked grammar (POC ADR-0014):
 *   - before_*: errorPolicy default `abort` — handler failure throws
 *     a `DiagnosticError` and the chokepoint decorator cancels the
 *     mutation.
 *   - after_*: errorPolicy default `continue` — handler failure is
 *     logged via `console.error` and swallowed; the mutation already
 *     succeeded. Author override (`errorPolicy: continue` on a
 *     before_* hook) is honored.
 *
 * `Trigger.errorPolicy: 'abort'` on after_* is parser-rejected, so the
 * use case never has to reconcile that combination.
 */
export class RunLifecycleHooksUseCase {
    triggers;
    proceduresByName;
    invoke;
    constructor(triggers, proceduresByName, invoke) {
        this.triggers = triggers;
        this.proceduresByName = proceduresByName;
        this.invoke = invoke;
    }
    async run(request) {
        const matching = this.triggers.forHook(request.schema, request.hook);
        if (matching.length === 0)
            return;
        const isBefore = request.hook.startsWith("before_");
        for (const trigger of matching) {
            await this.runOne(trigger, request, isBefore);
        }
    }
    async runOne(trigger, request, isBefore) {
        const procedure = this.proceduresByName.get(trigger.spec.target.procedure);
        if (!procedure) {
            // Boot validator should have caught this; defense-in-depth.
            throw new DiagnosticError(runtimeDiagnostic({
                code: "TRIGGER_TARGET_PROCEDURE_UNKNOWN",
                severity: "error",
                path: `manifest:Trigger/${trigger.metadata.name}#/spec/target/procedure`,
                value: trigger.spec.target.procedure,
                expected: "name of a declared Procedure",
                message: `Lifecycle Trigger '${trigger.metadata.name}' targets unknown Procedure '${trigger.spec.target.procedure}'.`,
            }));
        }
        const event = {
            hook: request.hook,
            schema: request.schema,
            entry: request.entry,
        };
        const ctxWithEvent = { ...request.ctx, event };
        const declared = trigger.spec.source.kind === "lifecycle"
            ? trigger.spec.source.errorPolicy
            : undefined;
        const policy = declared ?? (isBefore ? "abort" : "continue");
        const input = request.originalInput ?? request.entry?.data ?? {};
        let result;
        try {
            result = await this.invoke.execute({
                procedure,
                input,
                ctx: ctxWithEvent,
                pathPrefix: `lifecycle:Trigger/${trigger.metadata.name}->${procedure.metadata.name}`,
            });
        }
        catch (err) {
            if (policy === "abort")
                throw err;
            console.error(`[lifecycle] after_* hook ${trigger.metadata.name} threw`, err);
            return;
        }
        if (result.ok)
            return;
        if (policy === "abort") {
            throw new DiagnosticError(result.diagnostic);
        }
        console.error(`[lifecycle] hook ${trigger.metadata.name} returned diagnostic`, result.diagnostic);
    }
}
