// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/content/ArchiveUseCase.js
import { canTransition, DiagnosticError, } from "@aotterclam/clam-cms-spec";
import { unpublishCache, } from "./ContentPublishEffects.js";
import { illegalTransitionDiagnostic, notFoundDiagnostic, withConflictDiagnostic, } from "./diagnostics.js";
/**
 * `ArchiveUseCase` — flip an entry to `'archived'`. Per the simple-
 * lifecycle state machine, draft and published rows can both be
 * archived; archived rows go back to draft via `Unpublish`.
 */
export class ArchiveUseCase {
    entries;
    schemas;
    clock;
    effects;
    constructor(entries, schemas, clock, effects) {
        this.entries = entries;
        this.schemas = schemas;
        this.clock = clock;
        this.effects = effects;
    }
    async execute(request) {
        const opPath = `usecase/Archive/${request.id}`;
        const existing = await this.entries.get(request.id);
        if (!existing) {
            throw new DiagnosticError(notFoundDiagnostic(opPath, "<unknown>", request.id));
        }
        const schema = this.schemas.get(existing.collection);
        if (!canTransition(schema, existing.status, "archived")) {
            throw new DiagnosticError(illegalTransitionDiagnostic(opPath, existing.status, "archived"));
        }
        const archived = await withConflictDiagnostic(opPath, () => this.entries.archive({
            id: request.id,
            collection: existing.collection,
            expectedVersion: request.expectedVersion,
            now: this.clock.now(),
            hookContext: request.ctx,
            originalInput: request.originalInput,
        }));
        await unpublishCache(this.effects, archived.id);
        return archived;
    }
}
