// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/content/CreateDraftUseCase.js
import { DiagnosticError, } from "@aotterclam/clam-cms-spec";
import { projectAndStamp } from "../../domain/service/BuiltinProjector.js";
import { schemaUnknownDiagnostic, withConflictDiagnostic, } from "./diagnostics.js";
import { assertEntryWritable } from "./EntryWriteGuards.js";
/**
 * `CreateDraftUseCase` — create a new draft entry. Both `simple` and
 * `editorial` lifecycles start in `'draft'`; the difference shows up
 * at publish time.
 */
export class CreateDraftUseCase {
    entries;
    schemas;
    clock;
    idgen;
    siteConfig;
    constructor(entries, schemas, clock, idgen, siteConfig) {
        this.entries = entries;
        this.schemas = schemas;
        this.clock = clock;
        this.idgen = idgen;
        this.siteConfig = siteConfig;
    }
    async execute(request) {
        const opPath = `usecase/CreateDraft/${request.collection}`;
        const schema = this.schemas.get(request.collection);
        if (!schema) {
            throw new DiagnosticError(schemaUnknownDiagnostic(opPath, request.collection, [...this.schemas.keys()]));
        }
        const id = this.idgen.next();
        const now = this.clock.now();
        const ctx = authoringContext(request.ctx, request.authorId);
        const data = projectAndStamp({ schema, input: request.data, ctx, clockNow: now });
        await assertEntryWritable({
            opPath,
            entries: this.entries,
            schema,
            data,
            siteConfig: this.siteConfig,
        });
        return withConflictDiagnostic(opPath, () => this.entries.create({
            id,
            collection: request.collection,
            status: "draft",
            data,
            authorId: request.authorId,
            now,
            hookContext: ctx,
            originalInput: request.originalInput,
        }));
    }
}
function authoringContext(ctx, authorId) {
    if (ctx)
        return ctx;
    return {
        user: authorId ? { id: authorId } : null,
        staff: null,
        env: {},
    };
}
