// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/content/ListEntriesUseCase.js
import { DiagnosticError, } from "@aotterclam/clam-cms-spec";
import { clampLimit } from "../../domain/service/Pagination.js";
import { schemaUnknownDiagnostic } from "./diagnostics.js";
/**
 * `ListEntriesUseCase` — list entries in a collection, optionally
 * filtered by status. Asserts the collection is a declared Schema.
 *
 * Caller-supplied `limit` is clamped at this layer (the trust
 * boundary between MCP / admin transports and the chokepoint
 * `EntryRepository`).
 */
export class ListEntriesUseCase {
    entries;
    schemas;
    constructor(entries, schemas) {
        this.entries = entries;
        this.schemas = schemas;
    }
    async execute(request) {
        const opPath = `usecase/ListEntries/${request.collection}`;
        if (!this.schemas.has(request.collection)) {
            throw new DiagnosticError(schemaUnknownDiagnostic(opPath, request.collection, [...this.schemas.keys()]));
        }
        return this.entries.list({
            collection: request.collection,
            status: request.status,
            limit: clampLimit(request.limit),
        });
    }
}
