// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/content/UnpublishUseCase.js
import { DiagnosticError } from "@aotterclam/clam-cms-spec";
import { unpublishCache, } from "./ContentPublishEffects.js";
import { illegalTransitionDiagnostic, notFoundDiagnostic, withConflictDiagnostic, } from "./diagnostics.js";
/**
 * `UnpublishUseCase` — flip published / archived rows back to
 * `'draft'`. Used by the admin SPA's "edit a published entry" flow.
 */
export class UnpublishUseCase {
    entries;
    clock;
    effects;
    constructor(entries, clock, effects) {
        this.entries = entries;
        this.clock = clock;
        this.effects = effects;
    }
    async execute(request) {
        const opPath = `usecase/Unpublish/${request.id}`;
        const existing = await this.entries.get(request.id);
        if (!existing) {
            throw new DiagnosticError(notFoundDiagnostic(opPath, "<unknown>", request.id));
        }
        if (existing.status !== "published" && existing.status !== "archived") {
            throw new DiagnosticError(illegalTransitionDiagnostic(opPath, existing.status, "draft"));
        }
        const unpublished = await withConflictDiagnostic(opPath, () => this.entries.transitionStatus({
            id: request.id,
            collection: existing.collection,
            to: "draft",
            expectedStatus: existing.status,
            now: this.clock.now(),
            hookContext: request.ctx,
            originalInput: request.originalInput,
        }));
        await unpublishCache(this.effects, unpublished.id);
        return unpublished;
    }
}
