// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/content/UpdateDraftUseCase.js
import { DiagnosticError, runtimeDiagnostic, } from "@aotterclam/clam-cms-spec";
import { projectUpdateAndStamp } from "../../domain/service/BuiltinProjector.js";
import { notFoundDiagnostic, withConflictDiagnostic, } from "./diagnostics.js";
import { assertEntryWritable } from "./EntryWriteGuards.js";
/**
 * `UpdateDraftUseCase` — update a draft's data. Only entries in
 * `'draft'` status are editable — `published` / `archived` go via the
 * unpublish (back to draft) path first.
 */
export class UpdateDraftUseCase {
    entries;
    schemas;
    clock;
    siteConfig;
    constructor(entries, schemas, clock, siteConfig) {
        this.entries = entries;
        this.schemas = schemas;
        this.clock = clock;
        this.siteConfig = siteConfig;
    }
    async execute(request) {
        const opPath = `usecase/UpdateDraft/${request.id}`;
        const existing = await this.entries.get(request.id);
        if (!existing) {
            throw new DiagnosticError(notFoundDiagnostic(opPath, "<unknown>", request.id));
        }
        if (existing.status !== "draft") {
            throw new DiagnosticError(runtimeDiagnostic({
                code: "CONFLICT",
                severity: "error",
                path: opPath,
                value: existing.status,
                expected: "row.status === 'draft'",
                message: `Entry '${request.id}' is in status '${existing.status}'; only drafts are editable. Unpublish first.`,
            }));
        }
        const schema = this.schemas.get(existing.collection);
        if (!schema) {
            throw new DiagnosticError(runtimeDiagnostic({
                code: "BUILTIN_HANDLER_SCHEMA_UNKNOWN",
                severity: "error",
                path: `${opPath}#/collection`,
                value: existing.collection,
                expected: "name of a declared Schema",
                candidates: [...this.schemas.keys()],
                message: `Entry '${request.id}' belongs to unknown Schema '${existing.collection}'.`,
            }));
        }
        const now = this.clock.now();
        const ctx = authoringContext(request.ctx, existing.authorId);
        const data = projectUpdateAndStamp({
            schema,
            existing: existing.data,
            patch: request.data,
            ctx,
            clockNow: now,
        });
        await assertEntryWritable({
            opPath,
            entries: this.entries,
            schema,
            data,
            excludeId: existing.id,
            siteConfig: this.siteConfig,
        });
        return withConflictDiagnostic(opPath, () => this.entries.update({
            id: request.id,
            collection: existing.collection,
            expectedVersion: request.expectedVersion,
            data,
            now,
            hookContext: ctx,
            originalInput: request.originalInput,
        }));
    }
}
function authoringContext(ctx, authorId) {
    if (ctx)
        return ctx;
    return {
        user: authorId ? { id: authorId } : null,
        staff: null,
        env: {},
    };
}
