// @aotterclam/clam-cms-cloudflare@0.0.8-alpha/dist/mount/bootRuntimeOnce.js
import { createCmsRuntime, } from "@aotterclam/clam-cms-runtime";
export function createCmsRef(config) {
    const runtime = createCmsRuntime({
        manifests: config.manifests,
        handlers: config.handlers,
        templates: config.templates,
        siteDefaults: config.siteDefaults,
        publicPathResolver: config.publicPathResolver,
        mediaAllowSvg: config.mediaAllowSvg,
        mediaMaxBytes: config.mediaMaxBytes,
        db: config.bindings.db,
        kv: config.bindings.kv,
        assets: config.bindings.assets,
        mediaStorage: config.bindings.mediaStorage,
        deferredHookDispatcher: config.bindings.deferredHookDispatcher,
    });
    let booted = null;
    return {
        manifests: config.manifests,
        auth: config.auth,
        get() {
            if (booted)
                return booted;
            booted = runtime
                .bootInit()
                .then(() => runtime)
                .catch((err) => {
                booted = null;
                throw err;
            });
            return booted;
        },
    };
}
