// @aotterclam/clam-cms-cloudflare@0.0.8-alpha/dist/mount/mountPublicRoutes.js
import { entryHtmlKeyFromParts, entryMarkdownKeyFromParts, inferLocaleFromPath, isKnownLocale, listHtmlKey, llmsTxtKey, readEntryBySlug, serializeEntryAsMarkdown, toUrlLocale, } from "@aotterclam/clam-cms-runtime";
const HTML_NO_STORE = {
    "content-type": "text/html; charset=utf-8",
    "cache-control": "no-store",
};
const HTML_PUBLIC = {
    "content-type": "text/html; charset=utf-8",
};
const MD_PUBLIC = {
    "content-type": "text/markdown; charset=utf-8",
};
const TEXT_PUBLIC = {
    "content-type": "text/plain; charset=utf-8",
};
const SITEMAP_HEADERS = {
    "content-type": "application/xml; charset=utf-8",
    "cache-control": "public, max-age=300, s-maxage=300",
};
export function mountPublicRoutes(app, ref, options) {
    const liveDev = options.liveDev === true;
    const overrideIndex = buildOverrideIndex(options.slugOverrides ?? []);
    // Literal root paths register BEFORE any param-catch-all routes —
    // Hono's trie matches `/llms.txt` against `/:locale` with
    // `:locale = "llms.txt"` if the literal route registers later.
    app.get("/llms.txt", async () => {
        const runtime = await ref.get();
        return readKvText(runtime.kv, llmsTxtKey(""), TEXT_PUBLIC);
    });
    app.get("/sitemap.xml", async (c) => {
        const runtime = await ref.get();
        const site = await runtime.siteConfig.load();
        if (!runtime.publicPathResolver) {
            return new Response("sitemap unavailable: no publicPathResolver configured", {
                status: 500,
            });
        }
        const xml = await runtime.composeSitemap.execute({
            site,
            pathFor: (e) => runtime.publicPathResolver.forEntry(e),
        });
        return new Response(xml, { status: 200, headers: SITEMAP_HEADERS });
    });
    if (options.homeRenderer) {
        app.get("/", async (c) => {
            const runtime = await ref.get();
            const site = await runtime.siteConfig.load();
            const canonical = site.canonicalLocale ?? site.locales[0] ?? "en";
            return c.redirect(`/${toUrlLocale(canonical)}`);
        });
        app.get("/:locale", async (c) => {
            const runtime = await ref.get();
            const site = await runtime.siteConfig.load();
            const locale = canonicalLocaleParam(c.req.param("locale"), site);
            const ctx = buildCtx(c, runtime, site, locale ?? inferLocaleFromPath(c.req.path, site));
            if (locale === null)
                return options.notFoundRenderer(ctx);
            return options.homeRenderer(ctx);
        });
    }
    app.get("/:locale/llms.txt", async (c) => {
        const runtime = await ref.get();
        const site = await runtime.siteConfig.load();
        const locale = canonicalLocaleParam(c.req.param("locale"), site);
        if (locale === null)
            return new Response("not found", { status: 404, headers: TEXT_PUBLIC });
        return readKvText(runtime.kv, llmsTxtKey(locale), TEXT_PUBLIC);
    });
    for (const route of options.collectionRoutes) {
        mountCollection(app, ref, options, route, liveDev, overrideIndex);
    }
    app.notFound(async (c) => {
        const runtime = await ref.get();
        const site = await runtime.siteConfig.load();
        const locale = inferLocaleFromPath(c.req.path, site);
        return options.notFoundRenderer(buildCtx(c, runtime, site, locale));
    });
}
function mountCollection(app, ref, options, route, liveDev, overrides) {
    const segPath = route.segment ? `/${route.segment}` : "";
    if (route.listRoute) {
        app.get(`/:locale${segPath}`, async (c) => {
            const runtime = await ref.get();
            const site = await runtime.siteConfig.load();
            const locale = canonicalLocaleParam(c.req.param("locale"), site);
            const ctx = buildCtx(c, runtime, site, locale ?? inferLocaleFromPath(c.req.path, site));
            const notFound = () => options.notFoundRenderer(ctx);
            if (locale === null)
                return notFound();
            if (liveDev) {
                const html = await runtime.renderListLive.execute({
                    collection: route.collection,
                    locale,
                    site,
                });
                if (html === null)
                    return notFound();
                return new Response(html, { status: 200, headers: HTML_NO_STORE });
            }
            return readThroughCache(runtime.kv, listHtmlKey(route.collection, locale), HTML_PUBLIC, async () => runtime.renderListLive.execute({
                collection: route.collection,
                locale,
                site,
            }), notFound);
        });
    }
    // Register the `.md` mirror BEFORE the bare `:slug` entry route —
    // Hono matches in registration order, so without this the entry
    // route swallows `slug = "foo.md"` and 404s on KV lookup.
    if (route.markdownMirror !== false) {
        // `[^/]+\\.md` (not `.+\\.md`) so a malicious crawler can't
        // squat sub-paths like `/en/posts/long/random.md` and burn KV
        // reads — the `:slug` group stays single-segment.
        app.get(`/:locale${segPath}/:slug{[^/]+\\.md}`, async (c) => {
            const runtime = await ref.get();
            const site = await runtime.siteConfig.load();
            const locale = canonicalLocaleParam(c.req.param("locale"), site);
            const slugParam = c.req.param("slug") ?? "";
            const slug = slugParam.endsWith(".md") ? slugParam.slice(0, -3) : slugParam;
            const notFound = () => new Response("not found", { status: 404, headers: TEXT_PUBLIC });
            if (locale === null)
                return notFound();
            const key = entryMarkdownKeyFromParts(route.collection, locale, slug);
            return readThroughCache(runtime.kv, key, MD_PUBLIC, async () => {
                const entry = await readEntryBySlug(runtime.db, {
                    collection: route.collection,
                    slug,
                    locale,
                    status: "published",
                });
                if (!entry)
                    return null;
                return serializeEntryAsMarkdown(entry);
            }, notFound);
        });
    }
    app.get(`/:locale${segPath}/:slug`, async (c) => {
        const runtime = await ref.get();
        const site = await runtime.siteConfig.load();
        const locale = canonicalLocaleParam(c.req.param("locale"), site);
        const slug = c.req.param("slug");
        const ctx = buildCtx(c, runtime, site, locale ?? inferLocaleFromPath(c.req.path, site));
        const notFound = () => options.notFoundRenderer(ctx);
        if (locale === null)
            return notFound();
        const override = overrides.get(overrideKey(route.collection, slug));
        if (override)
            return override.render(ctx);
        if (c.req.query("preview") === "1") {
            const html = await runtime.previewEntry.execute({
                collection: route.collection,
                slug,
                locale,
                site,
            });
            if (html === null)
                return notFound();
            return new Response(html, { status: 200, headers: HTML_NO_STORE });
        }
        if (liveDev) {
            const html = await runtime.renderEntryLive.execute({
                collection: route.collection,
                slug,
                locale,
                site,
            });
            if (html === null)
                return notFound();
            return new Response(html, { status: 200, headers: HTML_NO_STORE });
        }
        const key = entryHtmlKeyFromParts(route.collection, locale, slug);
        return readThroughCache(runtime.kv, key, HTML_PUBLIC, async () => {
            const html = await runtime.renderEntryLive.execute({
                collection: route.collection,
                slug,
                locale,
                site,
            });
            return html;
        }, notFound);
    });
    if (route.homeSlug && options.homeRenderer == null) {
        // Without a homeRenderer the `homeSlug` collapse is a noop —
        // there's no `/{locale}` route to serve. Surface as a console
        // warning at boot rather than silently misconfiguring.
        console.warn(`[clam-cms] collectionRoute "${route.collection}" declares homeSlug="${route.homeSlug}" ` +
            `but no homeRenderer was passed to mountPublicRoutes — /{locale} will 404.`);
    }
}
async function readKvText(kv, key, headers) {
    const body = await kv.get(key);
    if (body === null) {
        return new Response("not found", { status: 404, headers: TEXT_PUBLIC });
    }
    return new Response(body, { status: 200, headers });
}
async function readThroughCache(kv, key, headers, populate, fallback) {
    const cached = await kv.get(key);
    if (cached !== null)
        return new Response(cached, { status: 200, headers });
    const rendered = await populate();
    if (rendered === null)
        return fallback();
    await kv.put(key, rendered);
    return new Response(rendered, { status: 200, headers });
}
function buildCtx(c, runtime, site, locale) {
    return { c, runtime, site, locale };
}
function canonicalLocaleParam(locale, site) {
    if (!isKnownLocale(locale, site))
        return null;
    return inferLocaleFromPath(`/${locale}`, site);
}
function buildOverrideIndex(overrides) {
    const map = new Map();
    for (const o of overrides)
        map.set(overrideKey(o.collection, o.slug), o);
    return map;
}
function overrideKey(collection, slug) {
    return `${collection}\u0000${slug}`;
}
