// @aotterclam/clam-cms-cloudflare@0.0.8-alpha/dist/mount/mountServerEndpoints.js
import { DiagnosticError, HTTP_STATUS_BY_CODE, MCP_HINT_KEYWORD, VIEW_PARAMS_RESERVED, isMediaMcpHint, redactForWire, runtimeDiagnostic, } from "@aotterclam/clam-cms-spec";
import { ViewParamCoercionError, coerceViewParams, matchPath, } from "@aotterclam/clam-cms-runtime";
import { indexHtml } from "@aotterclam/clam-cms-admin-ui";
import { ADMIN_ROLE_SET } from "../auth/createAuth.js";
import { AOTTER_CLAM_FAVICON_SVG } from "../assets/aotterClamFavicon.js";
const [PAGE_PARAM, SHOW_PARAM] = VIEW_PARAMS_RESERVED;
/** Mount HTTP Triggers + Views + the Better Auth admin surface. */
export function mountServerEndpoints(app, ref) {
    for (const t of ref.manifests) {
        if (t.kind !== "Trigger")
            continue;
        const source = t.spec.source;
        if (source.kind !== "http")
            continue;
        const { method, path } = source;
        const honoPath = openApiToHono(path);
        const triggerName = t.metadata.name;
        app.on(method, honoPath, async (c) => {
            const runtime = await ref.get();
            const waitUntil = readWaitUntil(c);
            return handleHttpTrigger(c.req.raw, runtime, ref.auth, triggerName, path, waitUntil);
        });
    }
    for (const v of ref.manifests) {
        if (v.kind !== "View")
            continue;
        const viewName = v.metadata.name;
        app.get(`/api/views/${viewName}`, async (c) => {
            const runtime = await ref.get();
            return handleViewRequest(c.req.raw, runtime, viewName);
        });
    }
    mountAdminBetterAuth(app, ref, ref.auth);
}
function mountAdminBetterAuth(app, ref, auth) {
    const spa = () => new Response(indexHtml, {
        headers: { "content-type": "text/html; charset=utf-8" },
    });
    app.get("/favicon.svg", () => new Response(AOTTER_CLAM_FAVICON_SVG, {
        headers: {
            "content-type": "image/svg+xml; charset=utf-8",
            "cache-control": "public, max-age=86400",
        },
    }));
    for (const path of [
        "/.well-known/oauth-authorization-server",
        "/.well-known/oauth-protected-resource",
    ]) {
        app.get(path, async (c) => {
            const url = new URL(c.req.url);
            url.pathname = `/api/auth${url.pathname}`;
            const res = await auth.handler(new Request(url.toString(), { method: "GET", headers: c.req.raw.headers }));
            if (!res.ok)
                return res;
            const origin = new URL(c.req.url).origin;
            const body = await res.json();
            if (path !== "/.well-known/oauth-authorization-server") {
                return Response.json({
                    ...body,
                    logo_uri: `${origin}/favicon.svg`,
                });
            }
            return Response.json({
                ...body,
                logo_uri: `${origin}/favicon.svg`,
                scopes_supported: [
                    "openid",
                    "profile",
                    "email",
                    "offline_access",
                    "mcp:read",
                    "mcp:staff",
                ],
            });
        });
    }
    for (const path of [
        "/admin",
        "/admin/",
        "/admin/sign-in",
        "/admin/c/:collection",
        "/admin/approvals",
        "/admin/preferences",
        "/admin/settings",
    ]) {
        app.get(path, spa);
    }
    // Pre-derive the collections projection — `ref.manifests` is
    // immutable post-boot, so the filter / Set / mediaFields work doesn't
    // need to repeat per request.
    const schemas = ref.manifests.filter((m) => m.kind === "Schema");
    const translatedParents = new Set();
    for (const s of schemas) {
        if (s.spec.translates)
            translatedParents.add(s.spec.translates.parent);
    }
    const collections = schemas
        .filter((s) => !s.spec.translates)
        .map((s) => ({
        name: s.metadata.name,
        title: s.spec.title,
        description: s.spec.description ?? null,
        lifecycle: s.spec.lifecycle ?? "simple",
        hasTranslations: translatedParents.has(s.metadata.name),
        mediaFields: mediaFieldsForCollection(s, schemas),
    }));
    const guarded = (method, path, body) => {
        app.on(method.toUpperCase(), path, async (c) => {
            const gate = await readAdminGate(c, auth);
            if (gate.kind === "unauth")
                return adminUnauthenticated(c, path);
            if (gate.kind === "forbidden")
                return adminNotStaff(c, path, gate.login);
            return body(c, gate);
        });
    };
    guarded("get", "/admin/api/me", (_c, gate) => jsonResponse(200, { login: gate.login, role: gate.role, userId: gate.userId }));
    guarded("get", "/admin/api/collections", () => jsonResponse(200, { collections }));
    guarded("get", "/admin/api/site", async (c) => {
        const runtime = await ref.get();
        const site = await runtime.siteConfig.load();
        const url = new URL(c.req.url);
        return jsonResponse(200, {
            ...site,
            publicUrl: site.origin || url.origin,
            mcpUrl: `${url.origin}/staff/mcp`,
            staffMcpUrl: `${url.origin}/staff/mcp`,
            userMcpUrl: `${url.origin}/mcp`,
        });
    });
    guarded("get", "/admin/api/entries", async (c) => {
        const collection = c.req.query("collection");
        if (!collection) {
            return jsonResponse(400, {
                ok: false,
                diagnostic: runtimeDiagnostic({
                    code: "INPUT_VALIDATION_FAILED",
                    severity: "error",
                    path: "GET /admin/api/entries",
                    expected: "?collection=<name> query parameter",
                    message: "Missing `collection` query parameter.",
                }),
            });
        }
        const runtime = await ref.get();
        const rows = await runtime.listEntries.execute({
            collection,
            status: c.req.query("status"),
        });
        const items = rows.map((row) => ({
            id: row.id,
            collection: row.collection,
            locale: row.locale ?? null,
            status: row.status,
            version: row.version,
            title: row.data.title,
            updated_at: row.updatedAt,
        }));
        return jsonResponse(200, { items, next_cursor: null });
    });
    // Three-step direct-upload flow: POST /uploads (capability) → caller
    // PUTs directly to R2 S3 (Worker bypassed) → POST /uploads/:id/commit.
    const MEDIA_UPLOADS_PATH = "/admin/api/media/uploads";
    const MEDIA_COMMIT_PATH = "/admin/api/media/uploads/:uploadId/commit";
    guarded("post", MEDIA_UPLOADS_PATH, async (c) => {
        const runtime = await ref.get();
        const media = runtime.media;
        if (!media)
            return mediaNotConfiguredResponse(`POST ${MEDIA_UPLOADS_PATH}`);
        const body = (await c.req.raw.json().catch(() => ({})));
        if (typeof body.filename !== "string" || typeof body.mimeType !== "string") {
            return jsonResponse(400, {
                ok: false,
                diagnostic: runtimeDiagnostic({
                    code: "INPUT_VALIDATION_FAILED",
                    severity: "error",
                    path: `POST ${MEDIA_UPLOADS_PATH}`,
                    expected: "{ filename: string, mimeType: string, byteSize?: number }",
                }),
            });
        }
        const { filename, mimeType } = body;
        return runUseCase(`POST ${MEDIA_UPLOADS_PATH}`, () => media.createUpload.execute({
            filename,
            mimeType,
            byteSize: typeof body.byteSize === "number" ? body.byteSize : undefined,
            alt: typeof body.alt === "string" ? body.alt : undefined,
            caption: typeof body.caption === "string" ? body.caption : undefined,
            purpose: typeof body.purpose === "string" ? body.purpose : undefined,
        }));
    });
    guarded("post", MEDIA_COMMIT_PATH, async (c) => {
        const runtime = await ref.get();
        const media = runtime.media;
        if (!media)
            return mediaNotConfiguredResponse(`POST ${MEDIA_COMMIT_PATH}`);
        // Hono only invokes this handler when the route matched, so the
        // path param is always present at runtime.
        const uploadId = c.req.param("uploadId");
        const body = (await c.req.raw.json().catch(() => ({})));
        return runUseCase(`POST ${MEDIA_COMMIT_PATH}`, () => media.commitUpload.execute({
            uploadId,
            alt: typeof body.alt === "string" ? body.alt : undefined,
            caption: typeof body.caption === "string" ? body.caption : undefined,
            checksum: typeof body.checksum === "string" ? body.checksum : undefined,
        }));
    });
}
async function readAdminGate(c, auth) {
    const session = await auth.getSession(c.req.raw);
    if (!session)
        return { kind: "unauth" };
    const role = session.user.role ?? null;
    const login = session.user.githubLogin ?? null;
    if (!role || !ADMIN_ROLE_SET.has(role)) {
        return { kind: "forbidden", login };
    }
    return {
        kind: "ok",
        userId: session.user.id,
        login,
        role: role,
    };
}
async function handleHttpTrigger(req, runtime, auth, triggerName, triggerPath, waitUntil) {
    const trigger = runtime.triggersByName.get(triggerName);
    if (!trigger) {
        return jsonError({ status: 500, code: "INTERNAL_ERROR", message: `Trigger '${triggerName}' missing post-boot.` });
    }
    const procName = trigger.spec.target.procedure;
    const procedure = runtime.proceduresByName.get(procName);
    if (!procedure) {
        return jsonError({ status: 500, code: "INTERNAL_ERROR", message: `Procedure '${procName}' missing post-boot.` });
    }
    const url = new URL(req.url);
    const params = matchPath(triggerPath, url.pathname) ?? {};
    const body = await readBody(req);
    // Spread order matters: URL path params are authoritative for the
    // resource identifier (a `DELETE /entries/{id}` body MUST NOT spoof
    // `id`). Body fields fill in non-path inputs only.
    const input = { ...body, ...params };
    const ctx = await buildHandlerContext(req, auth, waitUntil);
    const result = await runtime.invokeProcedure.execute({
        procedure,
        input,
        ctx,
        pathPrefix: `${req.method} ${triggerPath}`,
    });
    if (result.ok) {
        return jsonResponse(200, { ok: true, data: result.data });
    }
    const status = HTTP_STATUS_BY_CODE[result.diagnostic.code] ?? 500;
    return jsonResponse(status, { ok: false, diagnostic: result.diagnostic });
}
async function handleViewRequest(req, runtime, viewName) {
    const view = runtime.viewsByName.get(viewName);
    if (!view) {
        return jsonError({ status: 500, code: "INTERNAL_ERROR", message: `View '${viewName}' missing post-boot.` });
    }
    const url = new URL(req.url);
    const viewPath = `GET /api/views/${viewName}`;
    const page = parsePositiveInt(url.searchParams.get(PAGE_PARAM));
    const show = parsePositiveInt(url.searchParams.get(SHOW_PARAM));
    let params;
    try {
        params = coerceViewParams(view, url.searchParams);
    }
    catch (err) {
        if (err instanceof ViewParamCoercionError) {
            return jsonResponse(400, {
                ok: false,
                diagnostic: runtimeDiagnostic({
                    code: "INPUT_VALIDATION_FAILED",
                    severity: "error",
                    path: viewPath,
                    expected: "query string conforms to View.spec.params",
                    message: err.message,
                }),
            });
        }
        throw err;
    }
    const result = await runtime.executeView.execute({
        view,
        pathPrefix: viewPath,
        options: { params, page, show },
    });
    if (result.ok) {
        return jsonResponse(200, { ok: true, data: result.result });
    }
    const status = HTTP_STATUS_BY_CODE[result.diagnostic.code] ?? 500;
    return jsonResponse(status, { ok: false, diagnostic: result.diagnostic });
}
function parsePositiveInt(raw) {
    if (raw == null || raw === "")
        return undefined;
    const n = Number.parseInt(raw, 10);
    return Number.isFinite(n) && n > 0 ? n : undefined;
}
async function readBody(req) {
    if (req.method === "GET" || req.method === "DELETE" || req.method === "HEAD") {
        return {};
    }
    const ct = req.headers.get("content-type") ?? "";
    if (!ct.includes("json"))
        return {};
    try {
        const parsed = await req.json();
        return parsed ?? {};
    }
    catch {
        return {};
    }
}
async function buildHandlerContext(req, auth, waitUntil) {
    const wu = waitUntil ? { waitUntil } : {};
    const unauth = { user: null, staff: null, env: {}, ...wu };
    // Fast path: skip the D1 query for public Triggers (no bearer).
    if (!req.headers.get("authorization")?.startsWith("Bearer "))
        return unauth;
    const session = await auth.getMcpSession(req);
    if (!session)
        return unauth;
    const role = await auth.getUserRole(session.userId);
    const staff = role && ADMIN_ROLE_SET.has(role)
        ? { id: session.userId, role: role }
        : null;
    return { user: { id: session.userId }, staff, env: {}, ...wu };
}
function openApiToHono(path) {
    return path.replace(/\{([^}]+)\}/g, ":$1");
}
/**
 * Hono's `c.executionCtx` getter THROWS when the underlying request
 * was dispatched without an `ExecutionContext` (test harness via
 * `app.request(...)`, non-Workers runtimes). Treat the throw as
 * "no waitUntil available" and fall back to inline-await downstream.
 */
function readWaitUntil(c) {
    try {
        const ctx = c.executionCtx;
        return ctx.waitUntil.bind(ctx);
    }
    catch {
        return undefined;
    }
}
function jsonResponse(status, body) {
    return new Response(JSON.stringify(body), {
        status,
        headers: { "content-type": "application/json" },
    });
}
function mediaFieldsForCollection(schema, schemas) {
    const related = [
        schema,
        ...schemas.filter((s) => s.spec.translates?.parent === schema.metadata.name),
    ];
    const out = [];
    for (const s of related) {
        out.push(...mediaFieldsForSchema(s));
    }
    return out;
}
function mediaFieldsForSchema(schema) {
    const props = schema.spec.schema.properties ?? {};
    const out = [];
    for (const [name, prop] of Object.entries(props)) {
        if (typeof prop !== "object" || prop === null)
            continue;
        const hint = prop[MCP_HINT_KEYWORD];
        if (!isMediaMcpHint(hint))
            continue;
        out.push({ name, hint });
    }
    return out;
}
function adminUnauthenticated(c, path) {
    return jsonResponse(401, {
        ok: false,
        diagnostic: runtimeDiagnostic({
            code: "UNAUTHENTICATED",
            severity: "error",
            path: `${c.req.method} ${path}`,
            expected: "active session cookie",
            message: "Not signed in. Sign in via /admin/sign-in first.",
        }),
    });
}
// Distinct from UNAUTHENTICATED so the SPA can render an "access
// denied" view for users who DID sign in but lack a staff row,
// instead of bouncing them back to /admin/sign-in (which the OAuth
// re-auth then silently fast-forwards through, producing a visible
// 5-step redirect chain that looks like an infinite loop).
function adminNotStaff(c, path, login) {
    return jsonResponse(403, {
        ok: false,
        login,
        diagnostic: runtimeDiagnostic({
            code: "AUTH_DENIED",
            severity: "error",
            path: `${c.req.method} ${path}`,
            expected: "staff role for the signed-in user",
            message: "Signed in, but this account isn't on the admin staff list. Contact a site owner to be added.",
        }),
    });
}
function jsonError(args) {
    const diagnostic = {
        code: args.code,
        severity: "error",
        phase: "runtime",
        path: "mount/http",
        message: args.message,
    };
    return jsonResponse(args.status, { ok: false, diagnostic });
}
async function runUseCase(opPath, fn) {
    try {
        const result = await fn();
        return jsonResponse(200, result);
    }
    catch (e) {
        if (e instanceof DiagnosticError) {
            const status = HTTP_STATUS_BY_CODE[e.diagnostic.code] ?? 500;
            return jsonResponse(status, { ok: false, diagnostic: redactForWire(e.diagnostic) });
        }
        // Don't leak raw exception strings on the wire — R2 / D1 / aws4fetch
        // errors can carry bucket names, account IDs, or query fragments.
        console.error(`[runUseCase ${opPath}] unhandled error`, e);
        return jsonResponse(500, {
            ok: false,
            diagnostic: runtimeDiagnostic({
                code: "INTERNAL_ERROR",
                severity: "error",
                path: opPath,
                message: "An internal error occurred.",
            }),
        });
    }
}
function mediaNotConfiguredResponse(path) {
    return jsonResponse(501, {
        ok: false,
        diagnostic: runtimeDiagnostic({
            code: "MEDIA_NOT_CONFIGURED",
            severity: "error",
            path,
            message: "Media uploads are not enabled on this deployment. Bind a `mediaStorage` adapter in `createCmsRuntime` to enable.",
        }),
    });
}
