// @aotterclam/clam-cms-cloudflare@0.0.8-alpha/dist/mount/mountMcp.js
import { McpJsonRpcDispatcher } from "@aotterclam/clam-cms-runtime";
import { ADMIN_ROLE_SET } from "../auth/createAuth.js";
export function mountMcp(app, ref, options = {}) {
    const auth = ref.auth;
    const path = options.path ?? "/mcp";
    const surface = options.surface ?? "staff";
    const requiredScope = options.requiredScope ?? (surface === "staff" ? "mcp:staff" : "mcp:read");
    const metadataPath = `${path}/.well-known/oauth-protected-resource`;
    let dispatcher = null;
    app.get(metadataPath, (c) => {
        const url = new URL(c.req.url);
        return c.json({
            resource: `${url.origin}${path}`,
            authorization_servers: [url.origin],
            jwks_uri: `${url.origin}/api/auth/mcp/jwks`,
            logo_uri: `${url.origin}/favicon.svg`,
            scopes_supported: [
                "openid",
                "profile",
                "email",
                "offline_access",
                "mcp:read",
                "mcp:staff",
            ],
            bearer_methods_supported: ["header"],
            resource_signing_alg_values_supported: ["RS256", "none"],
        });
    });
    app.all(path, async (c) => {
        // Boot is independent of auth — fetch concurrently to save one
        // D1 round-trip on the hot path.
        const [session, runtime] = await Promise.all([
            auth.getMcpSession(c.req.raw),
            ref.get(),
        ]);
        if (!session)
            return new Response("unauthorized", unauthorized(c.req.url, requiredScope));
        if (!hasRequiredScope(session.scopes, requiredScope)) {
            return new Response("forbidden", forbidden(requiredScope));
        }
        const role = await auth.getUserRole(session.userId);
        if (surface === "staff" && (!role || !ADMIN_ROLE_SET.has(role))) {
            return new Response("forbidden", forbidden(requiredScope));
        }
        dispatcher ??= new McpJsonRpcDispatcher({
            listEntries: runtime.listEntries,
            getEntry: runtime.getEntry,
            createDraft: runtime.createDraft,
            updateDraft: runtime.updateDraft,
            requestPublish: runtime.requestPublish,
            unpublish: runtime.unpublish,
            archive: runtime.archive,
            deleteEntry: runtime.deleteEntry,
            executeView: runtime.executeView,
            media: runtime.media
                ? {
                    createUpload: runtime.media.createUpload,
                    commitUpload: runtime.media.commitUpload,
                }
                : undefined,
        }, [...runtime.schemasByName.values()], {
            surface,
            views: ref.manifests.filter((m) => m.kind === "View"),
        });
        return dispatcher.dispatch(c.req.raw, {
            userId: session.userId,
            staff: role && ADMIN_ROLE_SET.has(role)
                ? { userId: session.userId, role: role }
                : null,
        });
    });
}
function hasRequiredScope(scopes, required) {
    if (scopes.includes(required))
        return true;
    return required === "mcp:read" && scopes.includes("mcp:staff");
}
function unauthorized(requestUrl, requiredScope) {
    const url = new URL(requestUrl);
    const metadataPath = `${url.pathname}/.well-known/oauth-protected-resource`;
    return {
        status: 401,
        headers: {
            "www-authenticate": `Bearer realm="mcp", scope="${requiredScope}", resource_metadata="${url.origin}${metadataPath}"`,
            "access-control-expose-headers": "WWW-Authenticate",
        },
    };
}
function forbidden(requiredScope) {
    return {
        status: 403,
        headers: {
            "www-authenticate": `Bearer realm="mcp", error="insufficient_scope", scope="${requiredScope}"`,
            "access-control-expose-headers": "WWW-Authenticate",
        },
    };
}
