// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/render/ComposeLlmsTxtUseCase.js
import { readPublishedEntries } from "../../domain/service/PublishedEntries.js";
import { serializeLlmsTxt } from "../../domain/service/MarkdownSerializer.js";
/**
 * Compose a `/llms.txt` body from currently-published entries.
 *
 *   - locale: string  → entries with `data.locale === locale`
 *   - locale: null    → non-localized entries only (publish-pipeline
 *                       semantic; matches what HtmlPublishOrchestrator
 *                       writes to `llms:root` on a non-localized
 *                       publish). Consumers that want a cross-locale
 *                       aggregate at the root URL should iterate
 *                       site.locales themselves and concat.
 */
export class ComposeLlmsTxtUseCase {
    db;
    constructor(db) {
        this.db = db;
    }
    async execute(request) {
        const entries = await readPublishedEntries(this.db, { locale: request.locale });
        const grouped = groupByCollection(entries);
        return serializeLlmsTxt({
            site: request.site,
            locale: request.locale ?? "",
            entriesByCollection: grouped,
        });
    }
}
function groupByCollection(entries) {
    const out = new Map();
    for (const e of entries) {
        const arr = out.get(e.collection);
        if (arr)
            arr.push(e);
        else
            out.set(e.collection, [e]);
    }
    return out;
}
