// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/render/PreviewEntryUseCase.js
import { readEntryBySlug } from "../../domain/service/PublishedEntries.js";
import { renderEntryHtml } from "../../domain/service/HtmlRenderer.js";
import { injectPreviewBanner } from "../../domain/service/PreviewBanner.js";
import { composeSeoIfPathed, } from "./EntrySeoSupport.js";
/** Default fallback: prefer drafts, fall back to published, then
 *  archived. Authors typically open `?preview=1` to see WIP. */
const DEFAULT_PREVIEW_STATUS_ORDER = [
    "draft",
    "published",
    "archived",
];
/**
 * Render an entry with a preview banner. Walks `statusOrder` until a
 * matching row is found, renders via the registered template, then
 * injects the banner just inside `<body>`. Returns `null` when no
 * matching row exists or no template is registered.
 *
 * Composes SEO/AEO meta when the runtime was built with a
 * `publicPathResolver` — preview pages match the meta the published
 * version would carry, so authors see real shape during review.
 */
export class PreviewEntryUseCase {
    db;
    templates;
    paths;
    composeSeo;
    constructor(db, templates, paths, composeSeo) {
        this.db = db;
        this.templates = templates;
        this.paths = paths;
        this.composeSeo = composeSeo;
    }
    async execute(request) {
        const order = request.statusOrder ?? DEFAULT_PREVIEW_STATUS_ORDER;
        let entry = null;
        for (const status of order) {
            entry = await readEntryBySlug(this.db, {
                collection: request.collection,
                slug: request.slug,
                locale: request.locale,
                status,
            });
            if (entry)
                break;
        }
        if (!entry)
            return null;
        const seo = await composeSeoIfPathed(this.composeSeo, this.paths, entry, request.site);
        const html = renderEntryHtml({
            entry,
            site: request.site,
            templates: this.templates,
            seo,
        });
        if (html === null)
            return null;
        const banner = request.banner ??
            `<div class="preview-banner">Preview · ${entry.status} · ${request.slug}</div>`;
        return injectPreviewBanner(html, banner);
    }
}
