// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/render/RenderEntryLiveUseCase.js
import { readEntryBySlug } from "../../domain/service/PublishedEntries.js";
import { renderEntryHtml } from "../../domain/service/HtmlRenderer.js";
import { composeSeoIfPathed, } from "./EntrySeoSupport.js";
/**
 * Render a single entry from current DB state. Used by adapter live-
 * dev routes (`CLAM_LOCAL_DEV=1`) to bypass the KV cache and pick up
 * template / chrome edits without re-running the publish pipeline.
 *
 * When the runtime was built with a `publicPathResolver`, the SEO/AEO
 * meta block is composed and threaded into `EntryContext.seo` — so
 * live-rendered HTML carries the same meta KV-cached HTML does.
 *
 * Returns the full HTML document, or `null` when:
 *   - no entry matches `(collection, slug, locale, status)`
 *   - the collection has no registered entry template
 * Adapters map `null` to a 404 response.
 */
export class RenderEntryLiveUseCase {
    db;
    templates;
    paths;
    composeSeo;
    constructor(db, templates, paths, composeSeo) {
        this.db = db;
        this.templates = templates;
        this.paths = paths;
        this.composeSeo = composeSeo;
    }
    async execute(request) {
        const entry = await readEntryBySlug(this.db, {
            collection: request.collection,
            slug: request.slug,
            locale: request.locale,
            status: request.status ?? "published",
        });
        if (!entry)
            return null;
        const seo = await composeSeoIfPathed(this.composeSeo, this.paths, entry, request.site);
        return renderEntryHtml({
            entry,
            site: request.site,
            templates: this.templates,
            seo,
        });
    }
}
