// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/kernel/diagnostic.js
export const HTTP_STATUS_BY_CODE = {
    INPUT_VALIDATION_FAILED: 400,
    INVALID_LOCALE: 400,
    UNAUTHENTICATED: 401,
    AUTH_DENIED: 403,
    NOT_FOUND: 404,
    METHOD_NOT_ALLOWED: 405,
    CONFLICT: 409,
    TRANSLATES_PARENT_UNKNOWN: 409,
    HANDLER_NOT_REGISTERED: 500,
    DISPATCHER_NOT_BUILT: 501,
    INTERNAL_ERROR: 500,
    OUTPUT_VALIDATION_FAILED: 500,
    // Lifecycle before_* hook explicitly aborted the mutation. 409 is
    // the right fit (the hook rejected based on resource state — bot
    // check failed, rate limit hit, etc.); the structured diagnostic
    // includes the failing hook's name for surfacing to the caller.
    LIFECYCLE_HOOK_REJECTED: 409,
    MEDIA_NOT_CONFIGURED: 501,
    MEDIA_UPLOAD_EXPIRED: 410,
    MEDIA_OBJECT_NOT_FOUND: 409,
    MEDIA_MIME_REJECTED: 400,
    MEDIA_SIZE_EXCEEDED: 400,
    MEDIA_SVG_REJECTED: 400,
    MEDIA_CHECKSUM_MISMATCH: 409,
};
/** Resolve a Diagnostic's HTTP status for wire egress; unknown codes
 *  default to 500. Lives in the kernel so every adapter shares one
 *  mapping. */
export function httpStatusFor(d) {
    return HTTP_STATUS_BY_CODE[d.code] ?? 500;
}
/**
 * Build a Diagnostic with a derived `message`. Call sites populate
 * the structured fields; `message` is generated from them so prose
 * cannot drift from structure.
 */
export function makeDiagnostic(input) {
    const { code, phase, severity, path, value, expected, candidates, suggestion } = input;
    let msg = input.message;
    if (!msg) {
        const parts = [`[${phase}/${code}] at ${path}`];
        if (expected)
            parts.push(`expected ${expected}`);
        if (value !== undefined)
            parts.push(`got ${formatValue(value)}`);
        if (suggestion)
            parts.push(`(did you mean ${suggestion}?)`);
        msg = parts.join("; ");
    }
    return { code, phase, severity, path, value, expected, candidates, suggestion, message: msg };
}
/** Phase-stamping helpers — equivalent to `makeDiagnostic({...input, phase})`
 *  but read cleaner at call sites.
 *
 *  `testDiagnostic` ships with no spec-side caller today — the test
 *  harness referenced in ADR-0007 is the planned consumer; `phase: "test"`
 *  is part of the public Diagnostic contract per ADR-0008 regardless of
 *  whether spec emits it directly. Don't drop the helper or the phase
 *  value before the harness lands. */
export const validateDiagnostic = (input) => makeDiagnostic({ ...input, phase: "validate" });
export const testDiagnostic = (input) => makeDiagnostic({ ...input, phase: "test" });
export const bootDiagnostic = (input) => makeDiagnostic({ ...input, phase: "boot" });
export const runtimeDiagnostic = (input) => makeDiagnostic({ ...input, phase: "runtime" });
function formatValue(v) {
    if (v === null)
        return "null";
    if (v === undefined)
        return "undefined";
    if (typeof v === "string")
        return JSON.stringify(v);
    if (typeof v === "number" || typeof v === "boolean")
        return String(v);
    try {
        const s = JSON.stringify(v);
        return s.length > 100 ? s.slice(0, 97) + "..." : s;
    }
    catch {
        return "[unserializable]";
    }
}
/**
 * Throwable carrier for one or more Diagnostics. Layers that surface
 * an error across a transport boundary (HTTP handler, MCP tool call,
 * render pipeline) throw this so the boundary catch can detect a
 * structured payload and emit it on the wire instead of falling back
 * to the INTERNAL_ERROR envelope reserved for genuinely unexpected
 * throws.
 */
export class DiagnosticError extends Error {
    diagnostics;
    constructor(diagnostic) {
        const list = Array.isArray(diagnostic) ? diagnostic : [diagnostic];
        const head = list[0];
        super(head ? head.message : "DiagnosticError");
        this.name = "DiagnosticError";
        this.diagnostics = list;
    }
    /** First diagnostic — convenience for single-diagnostic call sites. */
    get diagnostic() {
        const d = this.diagnostics[0];
        if (!d)
            throw new Error("DiagnosticError carries no diagnostics");
        return d;
    }
}
/**
 * Strip security-sensitive fields for wire egress. Per ADR-0008:
 * never expose `candidates` to untrusted callers — that leaks
 * schema information. Internal phases (validate / test / boot)
 * skip this redaction.
 */
export function redactForWire(d) {
    if (d.candidates === undefined)
        return d;
    const { candidates: _omit, ...rest } = d;
    return rest;
}
/**
 * Inverse of `redactForWire`: tolerantly parse a JSON string into a
 * Diagnostic, returning `null` on anything that isn't a Diagnostic
 * (non-JSON, missing required fields, etc.). Intended for consumers
 * receiving HTTP error bodies — runtime egress, MCP `error.data`,
 * any future SDK adapter.
 */
export function parseWireDiagnostic(text) {
    if (!text)
        return null;
    let raw;
    try {
        raw = JSON.parse(text);
    }
    catch {
        return null;
    }
    if (raw === null || typeof raw !== "object" || Array.isArray(raw))
        return null;
    const parsed = raw;
    if (typeof parsed["code"] !== "string")
        return null;
    if (typeof parsed["message"] !== "string")
        return null;
    if (typeof parsed["path"] !== "string")
        return null;
    const phase = parsed["phase"];
    if (phase !== "validate" && phase !== "test" && phase !== "boot" && phase !== "runtime") {
        return null;
    }
    const severity = parsed["severity"];
    if (severity !== "error" && severity !== "warning")
        return null;
    return parsed;
}
/**
 * Walk a JSON Pointer (RFC 6901) on a value. Returns `undefined` for
 * any miss. Best-effort — used to surface the offending value inside
 * a diagnostic, not for strict pointer dereferencing.
 */
export function readJsonPointer(root, jsonPointer) {
    if (!jsonPointer || jsonPointer === "/")
        return root;
    const parts = jsonPointer.split("/").slice(1).map(unescapePointer);
    let cur = root;
    for (const part of parts) {
        if (cur && typeof cur === "object") {
            cur = cur[part];
        }
        else {
            return undefined;
        }
    }
    return cur;
}
function unescapePointer(s) {
    return s.replace(/~1/g, "/").replace(/~0/g, "~");
}
