// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/usecase/EmitTypesUseCase.js
import { partitionManifests } from "../domain/service/ManifestParser.js";
import { RESERVED_ENTRY_COLUMNS, } from "../domain/model/ManifestGrammar.js";
/**
 * Generate TypeScript declarations from v0.1 manifests:
 *   - Schema → Entry_<name> data shape
 *   - Procedure → ProcInput_<name> / ProcOutput_<name>
 *   - View → ViewRow_<name> projection over the parent Schema
 *
 * Covers v0.1 JsonSchema features (string / integer / number /
 * boolean / object / array / enum / required / items). Skips $ref /
 * oneOf / anyOf / allOf / if-then-else — none in v0.1 grammar; they
 * fall through to `unknown`.
 *
 * Pure: no I/O. CLI handles stdout.
 */
const RESERVED_COLUMN_TYPES = {
    id: "string",
    status: `"draft" | "published" | "archived"`,
    version: "number",
    createdAt: "number",
    updatedAt: "number",
    authorId: "string | null",
};
export class EmitTypesUseCase {
    execute(request) {
        const { schemas, procedures, views } = partitionManifests(request.manifests);
        const schemaByName = new Map(schemas.map((s) => [s.metadata.name, s]));
        const out = [];
        out.push(`// Generated by \`clam-cms emit-types\`. Do not edit by hand.`);
        out.push("");
        out.push(`export namespace ${request.namespace} {`);
        out.push("");
        for (const s of schemas) {
            out.push(`  /** Entry data for Schema '${s.metadata.name}' */`);
            out.push(`  export interface Entry_${tsId(s.metadata.name)} ${renderType(s.spec.schema, "  ")}`);
            out.push("");
        }
        for (const p of procedures) {
            out.push(`  /** Procedure '${p.metadata.name}' input */`);
            out.push(`  export interface ProcInput_${tsId(p.metadata.name)} ${renderType(p.spec.input, "  ")}`);
            out.push("");
            out.push(`  /** Procedure '${p.metadata.name}' output */`);
            out.push(`  export interface ProcOutput_${tsId(p.metadata.name)} ${renderType(p.spec.output, "  ")}`);
            out.push("");
        }
        for (const v of views) {
            const parent = schemaByName.get(v.spec.from);
            out.push(`  /** Row shape returned by GET /api/views/${v.metadata.name} */`);
            if (!parent) {
                out.push(`  export type ViewRow_${tsId(v.metadata.name)} = unknown;`);
            }
            else {
                const props = (parent.spec.schema.properties ?? {});
                const fields = v.spec.fields ?? [...RESERVED_ENTRY_COLUMNS, ...Object.keys(props)];
                const reservedSet = new Set(RESERVED_ENTRY_COLUMNS);
                const rendered = [];
                for (const f of fields) {
                    if (reservedSet.has(f)) {
                        rendered.push(`    ${f}: ${RESERVED_COLUMN_TYPES[f]};`);
                    }
                    else if (props[f]) {
                        rendered.push(`    ${f}?: ${renderTypeInline(props[f])};`);
                    }
                    else {
                        rendered.push(`    ${f}?: unknown;`);
                    }
                }
                out.push(`  export interface ViewRow_${tsId(v.metadata.name)} {`);
                out.push(...rendered);
                out.push(`  }`);
            }
            out.push("");
        }
        out.push(`}`);
        return { source: out.join("\n") + "\n" };
    }
    static run(request) {
        return new EmitTypesUseCase().execute(request);
    }
}
function tsId(name) {
    return name.replace(/[^a-zA-Z0-9_$]+/g, "_");
}
function renderType(schema, indent) {
    if (schema.type === "object") {
        const props = (schema.properties ?? {});
        const required = new Set(schema.required ?? []);
        const lines = ["{"];
        for (const [name, sub] of Object.entries(props)) {
            const optional = required.has(name) ? "" : "?";
            lines.push(`${indent}  ${tsKey(name)}${optional}: ${renderTypeInline(sub)};`);
        }
        lines.push(`${indent}}`);
        return lines.join("\n");
    }
    return renderTypeInline(schema);
}
function renderTypeInline(schema) {
    const t = schema.type;
    if (Array.isArray(schema.enum) && schema.enum.length > 0) {
        return schema.enum
            .map((v) => (typeof v === "string" ? JSON.stringify(v) : String(v)))
            .join(" | ");
    }
    if (Array.isArray(t)) {
        return t.map((kind) => primitiveType(kind, schema)).join(" | ");
    }
    if (t === "object") {
        return renderType(schema, "");
    }
    if (t === "array") {
        const items = schema.items ? renderTypeInline(schema.items) : "unknown";
        return `${items}[]`;
    }
    return primitiveType(typeof t === "string" ? t : "unknown", schema);
}
function primitiveType(t, schema) {
    switch (t) {
        case "string":
            return "string";
        case "integer":
        case "number":
            return "number";
        case "boolean":
            return "boolean";
        case "null":
            return "null";
        default:
            return schema.nullable === true ? "unknown | null" : "unknown";
    }
}
function tsKey(name) {
    return /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(name) ? name : JSON.stringify(name);
}
