// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/usecase/IntrospectManifestsUseCase.js
import { partitionManifests } from "../domain/service/ManifestParser.js";
/**
 * Walk a parsed manifest set and project the derived shape per atom.
 * Pure: no I/O, no FS. The CLI is the only adapter today; future
 * surfaces (admin UI, MCP introspection tool) call this same use case.
 */
export class IntrospectManifestsUseCase {
    execute(request) {
        const partitioned = partitionManifests(request.manifests);
        const schemas = partitioned.schemas.map((s) => {
            const properties = s.spec.schema.properties ?? {};
            return {
                name: s.metadata.name,
                title: s.spec.title,
                localized: s.spec.localized ?? false,
                lifecycle: s.spec.lifecycle ?? "simple",
                translates: s.spec.translates ?? null,
                uniqueIndexes: s.spec.uniqueIndexes ?? [],
                properties: Object.keys(properties),
            };
        });
        const views = partitioned.views.map((v) => ({
            name: v.metadata.name,
            from: v.spec.from,
            params: v.spec.params ?? null,
            filter: v.spec.filter ?? null,
            orderBy: v.spec.orderBy ?? [],
            fields: v.spec.fields ?? null,
            limit: v.spec.limit ?? null,
            restPath: `/api/views/${v.metadata.name}`,
        }));
        const procedures = partitioned.procedures.map((p) => ({
            name: p.metadata.name,
            handler: p.spec.handler,
            auth: p.spec.requires?.auth ?? null,
            input: p.spec.input,
            output: p.spec.output,
        }));
        const triggers = partitioned.triggers.map((t) => ({
            name: t.metadata.name,
            source: t.spec.source,
            target: t.spec.target,
        }));
        return { schemas, views, procedures, triggers, parseErrors: request.parseErrors };
    }
    static run(request) {
        return new IntrospectManifestsUseCase().execute(request);
    }
}
