// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/usecase/ValidateManifestsUseCase.js
import { validateDiagnostic, } from "../kernel/diagnostic.js";
import { CLAM_BIND_VALUES, } from "../domain/model/ManifestGrammar.js";
import { partitionManifests } from "../domain/service/ManifestParser.js";
import { checkLocaleAndTranslates } from "../domain/service/CrossSchemaChecker.js";
import { bestMatch, manifestPath } from "../domain/service/ManifestPathDiagnoser.js";
/**
 * `ValidateManifestsUseCase` — Loop 1 of the SDK authoring contract
 * (ADR-0007 / authoring-contract.md). Pure: no DB, no network. The
 * structural parser (`ManifestParser`) catches single-manifest
 * envelope / shape / DRAFT-key errors. This use case catches everything
 * that requires looking at MULTIPLE manifests together (cross-refs,
 * duplicates, path collisions) plus Schema-aware checks.
 *
 * Diagnostics carry phase: "validate".
 *
 * Stateless and dependency-free; instantiate once and reuse, or call
 * directly via the static `run` helper if construction ceremony adds
 * no value.
 */
export class ValidateManifestsUseCase {
    execute(request) {
        const diags = [];
        const partitioned = partitionManifests(request.manifests);
        const schemasByName = byName(partitioned.schemas);
        const proceduresByName = byName(partitioned.procedures);
        diags.push(...checkDuplicates("Schema", partitioned.schemas, request.filePaths));
        diags.push(...checkDuplicates("View", partitioned.views, request.filePaths));
        diags.push(...checkDuplicates("Procedure", partitioned.procedures, request.filePaths));
        diags.push(...checkDuplicates("Trigger", partitioned.triggers, request.filePaths));
        for (const s of partitioned.schemas) {
            diags.push(...checkSchemaInternals(s, request.filePaths));
        }
        diags.push(...checkLocaleAndTranslates({
            schemas: partitioned.schemas,
            phase: "validate",
            siteLocales: request.siteLocales,
            filePaths: request.filePaths,
        }));
        for (const v of partitioned.views) {
            diags.push(...checkViewRefs(v, schemasByName, request.filePaths));
        }
        for (const p of partitioned.procedures) {
            diags.push(...checkProcedureAuth(p, request.filePaths));
            diags.push(...checkBuiltinHandler(p, schemasByName, request.filePaths));
        }
        diags.push(...checkTriggerRefs(partitioned.triggers, proceduresByName, request.filePaths, schemasByName));
        if (request.handlerSource !== undefined) {
            diags.push(...checkHandlerRefsInSource(partitioned.procedures, request.handlerSource, request.filePaths));
        }
        let errorCount = 0;
        let warningCount = 0;
        for (const d of diags) {
            if (d.severity === "error")
                errorCount++;
            else
                warningCount++;
        }
        return { diagnostics: diags, errorCount, warningCount };
    }
    /** Convenience static — equivalent to `new ValidateManifestsUseCase().execute(req)`. */
    static run(request) {
        return new ValidateManifestsUseCase().execute(request);
    }
}
function byName(arr) {
    const m = new Map();
    for (const x of arr)
        m.set(x.metadata.name, x);
    return m;
}
function checkDuplicates(kind, arr, filePaths) {
    const seen = new Map();
    const out = [];
    for (const m of arr) {
        const c = (seen.get(m.metadata.name) ?? 0) + 1;
        seen.set(m.metadata.name, c);
        if (c === 2) {
            out.push(validateDiagnostic({
                code: "DUPLICATE_NAME",
                severity: "error",
                path: manifestPath(kind, m.metadata.name, "/metadata/name", filePaths),
                value: m.metadata.name,
                expected: `metadata.name unique within kind ${kind}`,
                message: `Two ${kind} manifests share metadata.name '${m.metadata.name}'.`,
            }));
        }
    }
    return out;
}
function checkSchemaInternals(s, filePaths) {
    const out = [];
    const schema = s.spec.schema;
    const properties = schema.properties ?? {};
    const ui = s.spec.uniqueIndexes ?? [];
    ui.forEach((composite, ci) => {
        composite.forEach((field, fi) => {
            if (!(field in properties)) {
                out.push(validateDiagnostic({
                    code: "UNIQUE_INDEX_FIELD_UNKNOWN",
                    severity: "error",
                    path: manifestPath("Schema", s.metadata.name, `/spec/uniqueIndexes/${ci}/${fi}`, filePaths),
                    value: field,
                    expected: `name of a property declared in spec.schema.properties`,
                    candidates: Object.keys(properties),
                    suggestion: bestMatch(field, Object.keys(properties)),
                    message: `Schema '${s.metadata.name}' uniqueIndexes references unknown field '${field}'.`,
                }));
            }
        });
    });
    for (const [propName, propSpec] of Object.entries(properties)) {
        const ps = propSpec;
        if (!ps || typeof ps !== "object")
            continue;
        const bind = ps["x-clam-bind"];
        if (typeof bind !== "string")
            continue;
        if (!CLAM_BIND_VALUES.includes(bind)) {
            out.push(validateDiagnostic({
                code: "BIND_VALUE_NOT_IN_ENUM",
                severity: "error",
                path: manifestPath("Schema", s.metadata.name, `/spec/schema/properties/${propName}/x-clam-bind`, filePaths),
                value: bind,
                expected: `one of ${CLAM_BIND_VALUES.join(", ")}`,
                candidates: [...CLAM_BIND_VALUES],
                suggestion: bestMatch(bind, [...CLAM_BIND_VALUES]),
                message: `Schema '${s.metadata.name}' property '${propName}' has illegal x-clam-bind value.`,
            }));
        }
    }
    return out;
}
function checkViewRefs(v, schemasByName, filePaths) {
    const out = [];
    const fromName = v.spec.from;
    const schema = schemasByName.get(fromName);
    if (!schema) {
        out.push(validateDiagnostic({
            code: "VIEW_FROM_UNKNOWN_SCHEMA",
            severity: "error",
            path: manifestPath("View", v.metadata.name, "/spec/from", filePaths),
            value: fromName,
            expected: "name of a declared Schema",
            candidates: [...schemasByName.keys()],
            suggestion: bestMatch(fromName, [...schemasByName.keys()]),
            message: `View '${v.metadata.name}' references unknown Schema '${fromName}'.`,
        }));
        return out;
    }
    const props = schema.spec.schema.properties ?? {};
    const reserved = ["id", "status", "version", "createdAt", "updatedAt", "authorId"];
    const validFieldNames = new Set([...Object.keys(props), ...reserved]);
    if (v.spec.fields) {
        v.spec.fields.forEach((f, i) => {
            if (!validFieldNames.has(f)) {
                out.push(validateDiagnostic({
                    code: "VIEW_FIELD_NOT_IN_SCHEMA",
                    severity: "error",
                    path: manifestPath("View", v.metadata.name, `/spec/fields/${i}`, filePaths),
                    value: f,
                    expected: `property of Schema '${fromName}' or a reserved metadata field`,
                    candidates: [...validFieldNames].sort(),
                    suggestion: bestMatch(f, [...validFieldNames]),
                    message: `View '${v.metadata.name}' field '${f}' is not declared on Schema '${fromName}'.`,
                }));
            }
        });
    }
    if (v.spec.filter) {
        out.push(...checkFilterFields(v.spec.filter, validFieldNames, v.metadata.name, fromName, "/spec/filter", filePaths));
    }
    if (v.spec.orderBy) {
        v.spec.orderBy.forEach((o, i) => {
            if (!validFieldNames.has(o.field)) {
                out.push(validateDiagnostic({
                    code: "VIEW_FIELD_NOT_IN_SCHEMA",
                    severity: "error",
                    path: manifestPath("View", v.metadata.name, `/spec/orderBy/${i}/field`, filePaths),
                    value: o.field,
                    expected: `property of Schema '${fromName}' or a reserved metadata field`,
                    candidates: [...validFieldNames].sort(),
                    suggestion: bestMatch(o.field, [...validFieldNames]),
                    message: `View '${v.metadata.name}' orderBy references unknown field '${o.field}'.`,
                }));
            }
        });
    }
    return out;
}
function checkFilterFields(node, validFields, viewName, schemaName, jsonPointer, filePaths) {
    if ("eq" in node) {
        if (!validFields.has(node.eq.field)) {
            return [
                validateDiagnostic({
                    code: "VIEW_FILTER_FIELD_NOT_IN_SCHEMA",
                    severity: "error",
                    path: manifestPath("View", viewName, `${jsonPointer}/eq/field`, filePaths),
                    value: node.eq.field,
                    expected: `property of Schema '${schemaName}' or a reserved metadata field`,
                    candidates: [...validFields].sort(),
                    suggestion: bestMatch(node.eq.field, [...validFields]),
                    message: `View '${viewName}' filter references unknown field '${node.eq.field}'.`,
                }),
            ];
        }
        return [];
    }
    if ("and" in node) {
        return node.and.flatMap((c, i) => checkFilterFields(c, validFields, viewName, schemaName, `${jsonPointer}/and/${i}`, filePaths));
    }
    return node.or.flatMap((c, i) => checkFilterFields(c, validFields, viewName, schemaName, `${jsonPointer}/or/${i}`, filePaths));
}
function checkBuiltinHandler(p, schemasByName, filePaths) {
    const h = p.spec.handler;
    if (h.kind !== "builtin")
        return [];
    const out = [];
    const target = schemasByName.get(h.schema);
    if (!target) {
        out.push(validateDiagnostic({
            code: "BUILTIN_HANDLER_SCHEMA_UNKNOWN",
            severity: "error",
            path: manifestPath("Procedure", p.metadata.name, "/spec/handler/schema", filePaths),
            value: h.schema,
            expected: "name of a declared Schema",
            candidates: [...schemasByName.keys()],
            suggestion: bestMatch(h.schema, [...schemasByName.keys()]),
            message: `Procedure '${p.metadata.name}' has handler.kind: builtin / schema: '${h.schema}', but no Schema by that name is declared.`,
        }));
        return out;
    }
    // op: archive is editorial-only. Inline the lifecycle resolution
    // here — we need only the Schema.spec.lifecycle default, not the
    // full state-machine helpers.
    const lifecycle = target.spec.lifecycle ?? "simple";
    if (h.op === "archive" && lifecycle !== "editorial") {
        out.push(validateDiagnostic({
            code: "BUILTIN_HANDLER_SCHEMA_NOT_EDITORIAL",
            severity: "error",
            path: manifestPath("Procedure", p.metadata.name, "/spec/handler/op", filePaths),
            value: "archive",
            expected: `Schema '${h.schema}' to declare lifecycle: editorial (op: archive is editorial-only — see ADR-0011)`,
            message: `Procedure '${p.metadata.name}' uses op: archive on Schema '${h.schema}', but that Schema's lifecycle is 'simple'. Either set Schema.spec.lifecycle: editorial or use op: delete.`,
        }));
    }
    return out;
}
function checkProcedureAuth(p, filePaths) {
    const out = [];
    const all = p.spec.requires?.auth?.all;
    if (!all)
        return out;
    all.forEach((pred, i) => {
        if (!isValidPredicate(pred)) {
            out.push(validateDiagnostic({
                code: "AUTH_PREDICATE_NOT_IN_ENUM",
                severity: "error",
                path: manifestPath("Procedure", p.metadata.name, `/spec/requires/auth/all/${i}`, filePaths),
                value: pred,
                expected: "'ctx.user' or { 'ctx.staff': [<role>, ...] }",
                message: `Procedure '${p.metadata.name}' has an illegal auth predicate.`,
            }));
        }
    });
    return out;
}
function isValidPredicate(p) {
    if (p === "ctx.user")
        return true;
    if (typeof p === "object" &&
        p !== null &&
        !Array.isArray(p) &&
        "ctx.staff" in p) {
        const roles = p["ctx.staff"];
        if (Array.isArray(roles) && roles.length > 0 && roles.every((r) => typeof r === "string")) {
            return true;
        }
    }
    return false;
}
function checkTriggerRefs(triggers, proceduresByName, filePaths, schemasByName) {
    const out = [];
    const httpRoutes = new Map();
    for (const t of triggers) {
        const procName = t.spec.target.procedure;
        if (!proceduresByName.has(procName)) {
            out.push(validateDiagnostic({
                code: "TRIGGER_TARGET_PROCEDURE_UNKNOWN",
                severity: "error",
                path: manifestPath("Trigger", t.metadata.name, "/spec/target/procedure", filePaths),
                value: procName,
                expected: "name of a declared Procedure",
                candidates: [...proceduresByName.keys()],
                suggestion: bestMatch(procName, [...proceduresByName.keys()]),
                message: `Trigger '${t.metadata.name}' targets unknown Procedure '${procName}'.`,
            }));
        }
        if (t.spec.source.kind === "http") {
            const key = `${t.spec.source.method} ${t.spec.source.path}`;
            const prior = httpRoutes.get(key);
            if (prior) {
                out.push(validateDiagnostic({
                    code: "TRIGGER_PATH_COLLISION",
                    severity: "error",
                    path: manifestPath("Trigger", t.metadata.name, "/spec/source", filePaths),
                    value: key,
                    expected: `unique (method, path) across all http Triggers (also declared by '${prior}')`,
                    message: `Trigger '${t.metadata.name}' shares route ${key} with Trigger '${prior}'.`,
                }));
            }
            else {
                httpRoutes.set(key, t.metadata.name);
            }
        }
        if (t.spec.source.kind === "lifecycle" && schemasByName && !schemasByName.has(t.spec.source.schema)) {
            const schemaName = t.spec.source.schema;
            out.push(validateDiagnostic({
                code: "LIFECYCLE_SCHEMA_UNKNOWN",
                severity: "error",
                path: manifestPath("Trigger", t.metadata.name, "/spec/source/schema", filePaths),
                value: schemaName,
                expected: "name of a declared Schema",
                candidates: [...schemasByName.keys()],
                suggestion: bestMatch(schemaName, [...schemasByName.keys()]),
                message: `Trigger '${t.metadata.name}' watches unknown Schema '${schemaName}'.`,
            }));
        }
    }
    return out;
}
function checkHandlerRefsInSource(procedures, source, filePaths) {
    const out = [];
    for (const p of procedures) {
        if (p.spec.handler.kind !== "ref")
            continue;
        const ref = p.spec.handler.ref;
        const escaped = ref.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
        const re = new RegExp(`["'\`]${escaped}["'\`]`);
        if (!re.test(source)) {
            out.push(validateDiagnostic({
                code: "HANDLER_NOT_REGISTERED",
                severity: "warning",
                path: manifestPath("Procedure", p.metadata.name, "/spec/handler/ref", filePaths),
                value: ref,
                expected: `string literal '${ref}' to appear in handler source (e.g. registerHandler('${ref}', ...) or handlers: { '${ref}': ... })`,
                message: `Procedure '${p.metadata.name}' handler.ref '${ref}' was not found in any handler source file. The boot-time validator will hard-fail if it isn't registered at runtime.`,
            }));
        }
    }
    return out;
}
/**
 * Backwards-compatible function alias for the v0.1.0 import name.
 * Will be dropped once consumers migrate to `ValidateManifestsUseCase`.
 */
export function check(request) {
    return ValidateManifestsUseCase.run(request);
}
