// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/media/CommitMediaUploadUseCase.js
import { DiagnosticError } from "@aotterclam/clam-cms-spec";
import { mediaUploadExpiredDiagnostic, } from "./diagnostics.js";
import { DEFAULT_MAX_BYTES, PENDING_UPLOAD_KV_PREFIX, } from "./mediaAllowlist.js";
export class CommitMediaUploadUseCase {
    storage;
    kv;
    clock;
    opts;
    constructor(storage, kv, clock, opts = {}) {
        this.storage = storage;
        this.kv = kv;
        this.clock = clock;
        this.opts = opts;
    }
    async execute(request) {
        const opPath = "usecase/CommitMediaUpload";
        const kvKey = `${PENDING_UPLOAD_KV_PREFIX}${request.uploadId}`;
        const raw = await this.kv.get(kvKey);
        if (!raw) {
            throw new DiagnosticError(mediaUploadExpiredDiagnostic(opPath, request.uploadId));
        }
        const record = JSON.parse(raw);
        const maxBytes = this.opts.maxBytes ?? DEFAULT_MAX_BYTES;
        const asset = await this.storage.commitUpload({
            uploadId: request.uploadId,
            storageKey: record.storageKey,
            expectedMimeType: record.expectedMimeType,
            maxBytes,
            alt: request.alt,
            caption: request.caption,
            checksum: request.checksum,
            now: this.clock.now(),
        });
        // Best-effort cleanup. KV TTL covers us if delete fails.
        await this.kv.delete(kvKey).catch(() => undefined);
        return asset;
    }
}
