// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/media/CreateMediaUploadUseCase.js
import { DiagnosticError } from "@aotterclam/clam-cms-spec";
import { mediaMimeRejectedDiagnostic, mediaSizeExceededDiagnostic, mediaSvgRejectedDiagnostic, } from "./diagnostics.js";
import { DEFAULT_MAX_BYTES, MEDIA_SVG_MIME, PENDING_UPLOAD_KV_PREFIX, PENDING_UPLOAD_KV_TTL_SECONDS, UPLOAD_URL_TTL_SECONDS, isAllowedMime, } from "./mediaAllowlist.js";
export class CreateMediaUploadUseCase {
    storage;
    kv;
    clock;
    opts;
    constructor(storage, kv, clock, opts = {
        allowSvg: false,
    }) {
        this.storage = storage;
        this.kv = kv;
        this.clock = clock;
        this.opts = opts;
    }
    async execute(request) {
        const opPath = "usecase/CreateMediaUpload";
        const maxBytes = this.opts.maxBytes ?? DEFAULT_MAX_BYTES;
        if (request.mimeType === MEDIA_SVG_MIME && !this.opts.allowSvg) {
            throw new DiagnosticError(mediaSvgRejectedDiagnostic(opPath));
        }
        if (!isAllowedMime(request.mimeType, this.opts.allowSvg)) {
            throw new DiagnosticError(mediaMimeRejectedDiagnostic(opPath, request.mimeType));
        }
        if (request.byteSize !== undefined && request.byteSize > maxBytes) {
            throw new DiagnosticError(mediaSizeExceededDiagnostic(opPath, request.byteSize, maxBytes));
        }
        const now = this.clock.now();
        const expiresAt = now + UPLOAD_URL_TTL_SECONDS * 1000;
        const capability = await this.storage.createUpload({
            filename: request.filename,
            mimeType: request.mimeType,
            byteSize: request.byteSize,
            maxBytes,
            purpose: request.purpose,
            now,
            expiresAt,
        });
        const record = {
            storageKey: capability.storageKey,
            expectedMimeType: request.mimeType,
            expectedSize: request.byteSize,
            expiresAt: capability.expiresAt,
            createdAt: now,
        };
        await this.kv.put(`${PENDING_UPLOAD_KV_PREFIX}${capability.uploadId}`, JSON.stringify(record), { expirationTtl: PENDING_UPLOAD_KV_TTL_SECONDS });
        return {
            uploadId: capability.uploadId,
            uploadUrl: capability.uploadUrl,
            method: capability.method,
            requiredHeaders: capability.requiredHeaders,
            expiresAt: capability.expiresAt,
        };
    }
}
