// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/media/diagnostics.js
import { makeDiagnostic } from "@aotterclam/clam-cms-spec";
const PHASE = "runtime";
export function mediaNotConfiguredDiagnostic(opPath) {
    return makeDiagnostic({
        code: "MEDIA_NOT_CONFIGURED",
        phase: PHASE,
        severity: "error",
        path: opPath,
        expected: "media storage adapter bound at runtime",
        message: "Media uploads are not enabled on this deployment. Bind a `mediaStorage` adapter in `createCmsRuntime` to enable.",
    });
}
export function mediaMimeRejectedDiagnostic(opPath, mime) {
    return makeDiagnostic({
        code: "MEDIA_MIME_REJECTED",
        phase: PHASE,
        severity: "error",
        path: opPath,
        value: mime,
        expected: "one of: image/png, image/jpeg, image/webp, image/gif",
    });
}
export function mediaSvgRejectedDiagnostic(opPath) {
    return makeDiagnostic({
        code: "MEDIA_SVG_REJECTED",
        phase: PHASE,
        severity: "error",
        path: opPath,
        value: "image/svg+xml",
        expected: "non-SVG image (object storage does not sanitize SVG payloads)",
        suggestion: "set MEDIA_ALLOW_SVG=1 on the adapter if SVG is required",
    });
}
export function mediaSizeExceededDiagnostic(opPath, byteSize, maxBytes) {
    return makeDiagnostic({
        code: "MEDIA_SIZE_EXCEEDED",
        phase: PHASE,
        severity: "error",
        path: opPath,
        value: byteSize,
        expected: `byteSize <= ${maxBytes}`,
    });
}
export function mediaUploadExpiredDiagnostic(opPath, uploadId) {
    return makeDiagnostic({
        code: "MEDIA_UPLOAD_EXPIRED",
        phase: PHASE,
        severity: "error",
        path: opPath,
        value: uploadId,
        expected: "an active upload capability (TTL elapsed or never created)",
        suggestion: "call create_media_upload again to get a fresh capability",
    });
}
export function mediaObjectNotFoundDiagnostic(opPath, uploadId) {
    return makeDiagnostic({
        code: "MEDIA_OBJECT_NOT_FOUND",
        phase: PHASE,
        severity: "error",
        path: opPath,
        value: uploadId,
        expected: "an object PUT to the storage backend before commit",
        suggestion: "PUT the file to the upload URL before calling commit_media_upload",
    });
}
