// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/media/mediaAllowlist.js
/**
 * Closed mime allowlist for v0.1.x media uploads. Adapter authors
 * cannot extend this without a grammar-revise round.
 *
 * SVG is intentionally absent from the always-on set — object stores
 * don't sanitize and a passthrough SVG can carry inline scripts. SVG
 * uploads are only accepted when the adapter wires the use case with
 * `allowSvg: true`, typically from a `MEDIA_ALLOW_SVG=1` env opt-in.
 */
export const MEDIA_MIME_ALLOWLIST = [
    "image/png",
    "image/jpeg",
    "image/webp",
    "image/gif",
];
export const MEDIA_SVG_MIME = "image/svg+xml";
export function isAllowedMime(mime, allowSvg) {
    if (mime === MEDIA_SVG_MIME)
        return allowSvg;
    return MEDIA_MIME_ALLOWLIST.includes(mime);
}
/** Default upload byte ceiling: 25 MB. Large enough for most cover
 *  images and short-form social posts; small enough to keep R2 free
 *  tier viable. Adapters may pass a different value via the use-case
 *  ctor when they have stricter / looser policies. */
export const DEFAULT_MAX_BYTES = 25 * 1024 * 1024;
/** Upload-capability TTL: 15 minutes. Bearer URL is short-lived. */
export const UPLOAD_URL_TTL_SECONDS = 15 * 60;
/** KV record TTL: 1 hour. Outlives the upload URL so slow commits
 *  don't lose the mapping; orphan sweep cleans uncommitted objects
 *  separately. */
export const PENDING_UPLOAD_KV_TTL_SECONDS = 60 * 60;
export const PENDING_UPLOAD_KV_PREFIX = "media:pending:";
/** Map a mime type to a conventional file extension for object key
 *  construction. Extension is purely cosmetic in R2 — server-side
 *  serving never infers content-type from the key. */
export function extensionForMime(mime) {
    switch (mime) {
        case "image/png":
            return "png";
        case "image/jpeg":
            return "jpg";
        case "image/webp":
            return "webp";
        case "image/gif":
            return "gif";
        case MEDIA_SVG_MIME:
            return "svg";
        default:
            return "bin";
    }
}
