// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/infrastructure/cli/loadManifests.js
import { readFile, readdir, stat } from "node:fs/promises";
import { join, resolve } from "node:path";
import { cwd } from "node:process";
import { parseManifests, ManifestParseError, } from "../../domain/service/ManifestParser.js";
import { validateDiagnostic, } from "../../kernel/diagnostic.js";
export async function loadManifestsFromRoot(rootArg) {
    const root = resolve(cwd(), rootArg);
    const manifests = [];
    const parseErrors = [];
    const filePaths = new Map();
    let entries;
    try {
        entries = await collectYamlFiles(root);
    }
    catch (err) {
        parseErrors.push(validateDiagnostic({
            code: "MANIFEST_ROOT_NOT_FOUND",
            severity: "error",
            path: root,
            expected: "an existing directory containing *.yaml manifest files",
            message: `Could not read manifest root ${root}: ${err instanceof Error ? err.message : String(err)}`,
        }));
        return { manifests, parseErrors, filePaths };
    }
    for (const file of entries) {
        let text;
        try {
            text = await readFile(file, "utf8");
        }
        catch (err) {
            parseErrors.push(validateDiagnostic({
                code: "MANIFEST_READ_FAILED",
                severity: "error",
                path: file,
                message: `Failed to read ${file}: ${err instanceof Error ? err.message : String(err)}`,
            }));
            continue;
        }
        try {
            const parsed = parseManifests(text);
            parseErrors.push(...parsed.diagnostics);
            parsed.manifests.forEach((m, i) => {
                manifests.push(m);
                filePaths.set(`${m.kind}/${m.metadata.name}`, { file, docIndex: i });
            });
        }
        catch (err) {
            const idx = err instanceof ManifestParseError ? err.docIndex : undefined;
            const pointer = err instanceof ManifestParseError ? err.pointer : undefined;
            const path = idx != null
                ? pointer
                    ? `${file}#/${idx}${pointer}`
                    : `${file}#/${idx}`
                : file;
            parseErrors.push(validateDiagnostic({
                code: "INVALID_MANIFEST_ENVELOPE",
                severity: "error",
                path,
                message: err instanceof Error ? err.message : String(err),
            }));
        }
    }
    return { manifests, parseErrors, filePaths };
}
async function collectYamlFiles(root) {
    const out = [];
    async function walk(dir) {
        const items = await readdir(dir, { withFileTypes: true });
        for (const it of items) {
            const full = join(dir, it.name);
            if (it.isDirectory())
                await walk(full);
            else if (it.isFile() && (it.name.endsWith(".yaml") || it.name.endsWith(".yml"))) {
                out.push(full);
            }
        }
    }
    const s = await stat(root);
    if (!s.isDirectory()) {
        throw new Error(`${root} is not a directory`);
    }
    await walk(root);
    out.sort();
    return out;
}
