// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/domain/service/HtmlRenderer.js
/**
 * Pure render functions over the consumer-supplied template registry.
 * Used by:
 *   - `HtmlPublishOrchestrator` (publish-time → write KV)
 *   - request-time route handlers in adapters (live-render bypass,
 *     preview surface)
 *
 * Both call sites previously inlined the template-lookup + doctype
 * concatenation. Extracting here keeps the contract single-sourced
 * — adding a doctype mode, an OG-meta wrapper, or a per-collection
 * pre/post hook is one edit. No I/O; no DB; no env access.
 */
const DEFAULT_DOCTYPE = "<!doctype html>";
/** Returns the full HTML doc (doctype + body) or `null` if no entry
 *  template is registered for `entry.collection`. */
export function renderEntryHtml(args) {
    const tpl = args.templates.getEntryTemplate(args.entry.collection);
    if (!tpl)
        return null;
    return ((args.doctype ?? DEFAULT_DOCTYPE) +
        tpl({ entry: args.entry, site: args.site, seo: args.seo }));
}
/** Returns the full HTML doc or `null` if no list template is
 *  registered for `collection`. */
export function renderListHtml(args) {
    const tpl = args.templates.getListTemplate(args.collection);
    if (!tpl)
        return null;
    return ((args.doctype ?? DEFAULT_DOCTYPE) +
        tpl({
            collection: args.collection,
            locale: args.locale,
            entries: args.entries,
            site: args.site,
            seo: args.seo,
        }));
}
