// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/domain/service/PublishedEntries.js
export async function readPublishedEntries(db, filter = {}) {
    const conditions = ["status = 'published'"];
    const binds = [];
    if (filter.locale === null) {
        conditions.push(`json_extract(data, '$.locale') IS NULL`);
    }
    else if (typeof filter.locale === "string") {
        conditions.push(`json_extract(data, '$.locale') = ?`);
        binds.push(filter.locale);
    }
    if (filter.collection) {
        conditions.push("collection = ?");
        binds.push(filter.collection);
    }
    let sql = `SELECT id, collection, status, version, data, created_at, updated_at` +
        ` FROM entries WHERE ${conditions.join(" AND ")} ORDER BY updated_at DESC`;
    if (typeof filter.limit === "number" && Number.isFinite(filter.limit) && filter.limit > 0) {
        sql += ` LIMIT ${Math.floor(filter.limit)}`;
    }
    const rows = await db.prepare(sql).bind(...binds).all();
    return rows.map(rowToEntry);
}
/** Single-entry lookup by id. Used by the publish pipeline to load
 *  the entry it's about to render; lifted into the domain service so
 *  the orchestrator stops owning row-to-domain mapping. */
export async function readEntryById(db, id) {
    const row = await db
        .prepare(`SELECT id, collection, status, version, data, created_at, updated_at` +
        ` FROM entries WHERE id = ?`)
        .bind(id)
        .first();
    return row ? rowToEntry(row) : null;
}
export async function readEntryBySlug(db, args) {
    const conditions = ["collection = ?", `json_extract(data, '$.slug') = ?`];
    const binds = [args.collection, args.slug];
    if (args.locale === null) {
        conditions.push(`json_extract(data, '$.locale') IS NULL`);
    }
    else if (typeof args.locale === "string") {
        conditions.push(`json_extract(data, '$.locale') = ?`);
        binds.push(args.locale);
    }
    if (args.status) {
        conditions.push("status = ?");
        binds.push(args.status);
    }
    const sql = `SELECT id, collection, status, version, data, created_at, updated_at` +
        ` FROM entries WHERE ${conditions.join(" AND ")} ORDER BY updated_at DESC LIMIT 1`;
    const row = await db.prepare(sql).bind(...binds).first();
    return row ? rowToEntry(row) : null;
}
function rowToEntry(row) {
    const data = JSON.parse(row.data);
    const dataLocale = data["locale"];
    return {
        id: row.id,
        collection: row.collection,
        locale: typeof dataLocale === "string" ? dataLocale : undefined,
        status: row.status,
        version: row.version,
        data,
        createdAt: row.created_at,
        updatedAt: row.updated_at,
    };
}
