// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/domain/service/ViewSqlCompiler.js
import { DiagnosticError, isParamRef, runtimeDiagnostic, } from "@aotterclam/clam-cms-spec";
import { clampPage, clampShow } from "./Pagination.js";
// alias → SQL column. Aliases mirror RESERVED_ENTRY_COLUMNS from
// spec; SQL column shape (snake_case) is local to the storage layout.
// The compile-time check below ensures the alias set stays in sync —
// adding to spec without updating here is a type error.
const RESERVED_COLUMN = {
    id: "id",
    status: "status",
    version: "version",
    createdAt: "created_at",
    updatedAt: "updated_at",
    authorId: "author_id",
};
const _aliasCheck = RESERVED_COLUMN;
void _aliasCheck;
const DEFAULT_PROJECTION = Object.entries(RESERVED_COLUMN)
    .map(([alias, col]) => (alias === col ? col : `${col} AS ${alias}`))
    .join(", ");
export function compileView(view, options = {}) {
    const sqlParams = [];
    const selectExpr = buildSelect(view.spec.fields);
    const whereParts = ["collection = ?"];
    sqlParams.push(view.spec.from);
    if (view.spec.filter) {
        const filterSql = compileFilter(view.spec.filter, sqlParams, options.params ?? {});
        if (filterSql !== null)
            whereParts.push(`(${filterSql})`);
    }
    const where = `WHERE ${whereParts.join(" AND ")}`;
    const orderBy = buildOrderBy(view.spec.orderBy);
    const effectiveShow = clampShow(options.show, view.spec.limit);
    const effectivePage = clampPage(options.page);
    const offset = (effectivePage - 1) * effectiveShow;
    const sql = `SELECT ${selectExpr} FROM entries ${where}${orderBy} LIMIT ${effectiveShow} OFFSET ${offset}`;
    return { sql, params: sqlParams, effectivePage, effectiveShow };
}
function buildSelect(fields) {
    if (!fields || fields.length === 0)
        return DEFAULT_PROJECTION;
    return fields.map(fieldExpr).join(", ");
}
function fieldExpr(field) {
    const reserved = RESERVED_COLUMN[field];
    if (reserved) {
        return reserved === field ? reserved : `${reserved} AS ${field}`;
    }
    return `json_extract(data, '$.${escapeJsonKey(field)}') AS ${quoteIdent(field)}`;
}
function fieldRefExpr(field) {
    const reserved = RESERVED_COLUMN[field];
    if (reserved)
        return reserved;
    return `json_extract(data, '$.${escapeJsonKey(field)}')`;
}
/**
 * Returns `null` when the node should evaluate to TRUE (no constraint).
 * That happens when an `eq` references a param via `{ $param: <name> }`
 * and the resolved value is `undefined`. v0.1.0 parser enforces
 * required-only param refs, so this drop path is dead code today; it
 * lives here so v0.1.x "optional param ref" promotion is purely a
 * parser change. AND/OR fold over their children: any null child
 * drops; an AND/OR whose children all drop returns null itself.
 */
function compileFilter(node, sqlParams, paramValues) {
    if ("eq" in node) {
        const value = node.eq.value;
        if (isParamRef(value)) {
            const resolved = paramValues[value.$param];
            if (resolved === undefined)
                return null;
            sqlParams.push(resolved);
        }
        else {
            sqlParams.push(value);
        }
        return `${fieldRefExpr(node.eq.field)} = ?`;
    }
    if ("and" in node) {
        const parts = node.and
            .map((c) => compileFilter(c, sqlParams, paramValues))
            .filter((p) => p !== null);
        if (parts.length === 0)
            return null;
        return parts.map((p) => `(${p})`).join(" AND ");
    }
    const parts = node.or
        .map((c) => compileFilter(c, sqlParams, paramValues))
        .filter((p) => p !== null);
    if (parts.length === 0)
        return null;
    return parts.map((p) => `(${p})`).join(" OR ");
}
function buildOrderBy(orderBy) {
    if (!orderBy || orderBy.length === 0)
        return "";
    const parts = orderBy.map((o) => {
        const dir = (o.direction ?? "asc").toUpperCase();
        return `${fieldRefExpr(o.field)} ${dir}`;
    });
    return ` ORDER BY ${parts.join(", ")}`;
}
function escapeJsonKey(key) {
    if (key.includes("\\")) {
        throw new DiagnosticError(runtimeDiagnostic({
            code: "INTERNAL_ERROR",
            severity: "error",
            path: "compileView/escapeJsonKey",
            value: key,
            expected: "field name without backslashes (Schema validator gate)",
            message: `field name '${key}' contains a backslash; Schema validation should have caught this.`,
        }));
    }
    return key.replace(/'/g, "''");
}
function quoteIdent(name) {
    if (name.includes('"')) {
        throw new DiagnosticError(runtimeDiagnostic({
            code: "INTERNAL_ERROR",
            severity: "error",
            path: "compileView/quoteIdent",
            value: name,
            expected: "field name without double-quotes (Schema validator gate)",
            message: `field name '${name}' contains a double-quote; Schema validation should have caught this.`,
        }));
    }
    return `"${name}"`;
}
