// @aotterclam/clam-cms-spec@0.0.8-alpha/dist/domain/model/ManifestGrammar.js
/**
 * K8s-style manifest envelope, scoped to the cms group of the clam universe.
 *
 *   apiVersion: cms.clam.ai/v1
 *   kind: <Kind>
 *   metadata: { name }
 *   spec: { ... kind-specific ... }
 *
 * Sibling group `analytics.clam.ai/v1` lives in the parallel clam (OLAP)
 * project and is parsed there — the two systems share lineage and clam.ai
 * domain but not parsers.
 *
 * v0.1 grammar lock — see `docs/design-atoms.md` and ADR-0001
 * § "Future grammar discipline". DRAFT keys (policies, recursive views,
 * temporal predicates, quotas, projection triggers, mcp / cron / queue
 * Trigger source kinds) are intentionally absent from this file. They
 * land alongside their design pass; until then, the parser warns on
 * their use.
 *
 * v0.1.x extensions — see ADR-0001 § "What's DRAFT". `handler.kind: builtin`
 * and `Trigger.source.kind: lifecycle` are reserved as v0.1.x-committed
 * (will land before v0.2 once the lifecycle / editorial design pass
 * settles); the parser rejects them today with a fine-grained code so
 * authors who reach for them get a clear "not in v0.1.0 yet" message
 * rather than the generic DRAFT_KEY_USED.
 */
export const API_VERSION = "cms.clam.ai/v1";
export const MCP_HINTS = [
    "markdown",
    "richtext",
    "code",
    "media",
    "media-image",
    "media-video",
    "media-file",
];
export const MEDIA_MCP_HINTS = [
    "media",
    "media-image",
    "media-video",
    "media-file",
];
export function isMediaMcpHint(value) {
    return typeof value === "string" && MEDIA_MCP_HINTS.includes(value);
}
export const CLAM_REF_KEYWORD = "x-clam-ref";
export const MCP_HINT_KEYWORD = "x-mcp-hint";
export const CLAM_BIND_KEYWORD = "x-clam-bind";
export function isParamRef(v) {
    return (typeof v === "object" &&
        v !== null &&
        !Array.isArray(v) &&
        typeof v["$param"] === "string");
}
/** Reserved query-string names on the public View REST surface
 *  (`/api/views/<name>?...`). The runtime owns these for pagination; a
 *  View manifest that declares `params.<name>` for any of them is
 *  rejected at parse time (`VIEW_PARAMS_RESERVED_NAME`). */
export const VIEW_PARAMS_RESERVED = ["page", "show", "cursor"];
/** v0.1.x-committed builtin op vocabulary — see ADR-0001 § "What's DRAFT".
 *  Reserved for the lifecycle / editorial design pass; not implemented
 *  in v0.1.0. New entries require an explicit grammar-revise round. */
export const BUILTIN_OPS = ["create", "update", "upsert", "delete", "archive"];
/**
 * Staff role hierarchy. `users` is the base identity layer (runtime);
 * `staff` is the privilege overlay (one row per privileged user). A
 * user without a staff row is a regular site member with no admin
 * access.
 *
 *   owner       — full control, manages staff, manages settings
 *   editor      — publish, approve/reject, manage all entries
 *   contributor — create drafts, request publish, sees only own entries
 *
 * `StaffRole` lives in spec because the manifest grammar references
 * it directly: `requires.auth.all: [{ "ctx.staff": [<role>, ...] }]`
 * — the parser checks each role string is in `STAFF_ROLES` at boot.
 * The `Staff` runtime row shape (with grantedBy / grantedAt) lives in
 * `clam-cms-runtime`; only the closed enum is grammar.
 */
export const STAFF_ROLES = ["owner", "editor", "contributor"];
const ROLE_RANK = {
    owner: 3,
    editor: 2,
    contributor: 1,
};
export function meetsRole(actual, min) {
    return ROLE_RANK[actual] >= ROLE_RANK[min];
}
export function isStaffRole(s) {
    return STAFF_ROLES.includes(s);
}
/** v0.1.x-committed lifecycle hook vocabulary. The 6 SQL-92 hooks
 *  (before/after × create/update/delete) plus the editorial publish
 *  boundary (before/after publish). Not implemented in v0.1.0. New
 *  entries require an explicit grammar-revise round. */
export const LIFECYCLE_HOOKS = [
    "before_create",
    "after_create",
    "before_update",
    "after_update",
    "before_delete",
    "after_delete",
    "before_publish",
    "after_publish",
];
/** v0.1 closed enum for `x-clam-bind` Schema-property values. New entries
 *  require an explicit grammar-revise round (see ADR-0002). */
export const CLAM_BIND_VALUES = ["ctx.user", "ctx.staff", "now"];
/** Storage-row metadata columns reserved across every Schema. Used by
 *  the View SQL compiler (to project them as native columns rather
 *  than `json_extract`) and by the type emitter (to surface them on
 *  every Entry interface). Adding a new reserved column is a grammar
 *  revise — touch this constant + every consumer. */
export const RESERVED_ENTRY_COLUMNS = [
    "id",
    "status",
    "version",
    "createdAt",
    "updatedAt",
    "authorId",
];
