// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/infrastructure/mcp/McpJsonRpcDispatcher.js
import { DiagnosticError, redactForWire, } from "@aotterclam/clam-cms-spec";
import { mcpToolNameSegment } from "../../domain/service/McpToolNaming.js";
import { CREATE_DRAFT_PREFIX, QUERY_VIEW_PREFIX, UPDATE_DRAFT_PREFIX, buildMcpToolCatalog, extractCollectionSegment, } from "./McpToolCatalog.js";
import { jsonRpcError, jsonRpcOk, jsonRpcOkRaw, } from "./McpResponses.js";
export class McpJsonRpcDispatcher {
    useCases;
    schemas;
    options;
    catalog;
    catalogWireJson;
    /** segment → original `Schema.metadata.name`. Built once at
     *  construction; the per-collection routing path looks up the
     *  segment from the tool name and recovers the canonical
     *  collection name. */
    schemaBySegment;
    viewBySegment;
    constructor(useCases, schemas, options = {}) {
        this.useCases = useCases;
        this.schemas = schemas;
        this.options = options;
        this.catalog = buildMcpToolCatalog(schemas, {
            surface: options.surface ?? "staff",
            mediaEnabled: useCases.media !== undefined,
            views: options.views,
        });
        this.catalogWireJson = `{"tools":${JSON.stringify(this.catalog)}}`;
        const map = new Map();
        for (const s of schemas)
            map.set(mcpToolNameSegment(s.metadata.name), s.metadata.name);
        this.schemaBySegment = map;
        const views = new Map();
        for (const v of options.views ?? [])
            views.set(mcpToolNameSegment(v.metadata.name), v);
        this.viewBySegment = views;
    }
    async dispatch(req, auth) {
        if (req.method === "GET") {
            return new Response("MCP endpoint — POST JSON-RPC here.", { status: 200 });
        }
        if (req.method !== "POST") {
            return new Response("method not allowed", { status: 405 });
        }
        let body;
        try {
            body = (await req.json());
        }
        catch {
            return jsonRpcError(null, -32700, "parse error");
        }
        const { id = null, method, params } = body;
        switch (method) {
            case "initialize":
                return jsonRpcOk(id, {
                    protocolVersion: "2025-03-26",
                    capabilities: { tools: { listChanged: false } },
                    serverInfo: { name: "@aotterclam/clam-cms-runtime/mcp", version: "0.0.7-alpha" },
                });
            case "tools/list":
                return jsonRpcOkRaw(id, this.catalogWireJson);
            case "tools/call":
                return this.handleToolCall(id, params, auth);
            default:
                return jsonRpcError(id, -32601, `unknown method: ${method}`);
        }
    }
    async handleToolCall(reqId, params, auth) {
        const p = params;
        if (!p || typeof p.name !== "string") {
            return jsonRpcError(reqId, -32602, "missing tool name");
        }
        const args = (p.arguments ?? {});
        try {
            const result = await this.dispatchToolByName(p.name, args, auth);
            if (result === UNKNOWN_TOOL) {
                return jsonRpcError(reqId, -32601, `unknown tool: ${p.name}`);
            }
            if (result === MISSING_ARG) {
                return jsonRpcError(reqId, -32602, "missing required arg");
            }
            return jsonRpcOk(reqId, {
                content: [{ type: "text", text: JSON.stringify(result) }],
            });
        }
        catch (e) {
            if (e instanceof DiagnosticError) {
                return jsonRpcError(reqId, -32000, e.diagnostic.message, redactForWire(e.diagnostic));
            }
            // Don't leak raw exception strings to MCP clients — adapter
            // exceptions can carry binding / driver detail. Real cause goes
            // to server-side logs; the wire stays opaque.
            console.error("[McpJsonRpcDispatcher] unhandled tool-call error", e);
            return jsonRpcError(reqId, -32000, "Internal error.");
        }
    }
    async dispatchToolByName(name, args, auth) {
        if ((this.options.surface ?? "staff") === "public") {
            const viewSegment = extractCollectionSegment(name, QUERY_VIEW_PREFIX);
            if (!viewSegment)
                return UNKNOWN_TOOL;
            const view = this.viewBySegment.get(viewSegment);
            if (!view || !this.useCases.executeView)
                return UNKNOWN_TOOL;
            const result = await this.useCases.executeView.execute({
                view,
                options: {
                    params: stripViewReservedArgs(args),
                    page: typeof args["page"] === "number" ? args["page"] : undefined,
                    show: typeof args["show"] === "number" ? args["show"] : undefined,
                },
                pathPrefix: `MCP ${name}`,
            });
            return result;
        }
        switch (name) {
            case "list_entries": {
                const collection = args["collection"];
                if (typeof collection !== "string")
                    return MISSING_ARG;
                return this.useCases.listEntries.execute({
                    collection,
                    status: args["status"],
                    limit: typeof args["limit"] === "number" ? args["limit"] : undefined,
                });
            }
            case "get_entry": {
                const id = args["id"];
                if (typeof id !== "string")
                    return MISSING_ARG;
                return this.useCases.getEntry.execute({ id });
            }
            case "request_publish": {
                const id = args["id"];
                if (typeof id !== "string")
                    return MISSING_ARG;
                return this.useCases.requestPublish.execute({ id });
            }
            case "unpublish_entry": {
                const id = args["id"];
                if (typeof id !== "string")
                    return MISSING_ARG;
                return this.useCases.unpublish.execute({ id });
            }
            case "archive_entry": {
                const id = args["id"];
                const expected = args["expected_version"];
                if (typeof id !== "string" || typeof expected !== "number")
                    return MISSING_ARG;
                return this.useCases.archive.execute({ id, expectedVersion: expected });
            }
            case "create_media_upload": {
                if (!this.useCases.media)
                    return UNKNOWN_TOOL;
                const filename = args["filename"];
                const mimeType = args["mimeType"];
                if (typeof filename !== "string" || typeof mimeType !== "string")
                    return MISSING_ARG;
                return this.useCases.media.createUpload.execute({
                    filename,
                    mimeType,
                    byteSize: typeof args["byteSize"] === "number" ? args["byteSize"] : undefined,
                    alt: typeof args["alt"] === "string" ? args["alt"] : undefined,
                    caption: typeof args["caption"] === "string" ? args["caption"] : undefined,
                    purpose: typeof args["purpose"] === "string" ? args["purpose"] : undefined,
                });
            }
            case "commit_media_upload": {
                if (!this.useCases.media)
                    return UNKNOWN_TOOL;
                const uploadId = args["uploadId"];
                if (typeof uploadId !== "string")
                    return MISSING_ARG;
                return this.useCases.media.commitUpload.execute({
                    uploadId,
                    alt: typeof args["alt"] === "string" ? args["alt"] : undefined,
                    caption: typeof args["caption"] === "string" ? args["caption"] : undefined,
                    checksum: typeof args["checksum"] === "string" ? args["checksum"] : undefined,
                });
            }
            default: {
                // Per-collection authoring tools: `create_draft_<segment>` /
                // `update_draft_<segment>`. The agent sends Schema fields at
                // the top level; we rebuild `data` for the chokepoint.
                // `hookCtx` plumbs the authenticated MCP user into lifecycle
                // hook context so consumers can branch on `ctx.user` (e.g.
                // bypass captcha checks for authenticated agents).
                const hookCtx = {
                    user: { id: auth.userId },
                    staff: auth.staff ? { id: auth.staff.userId, role: auth.staff.role } : null,
                    env: {},
                };
                const createSegment = extractCollectionSegment(name, CREATE_DRAFT_PREFIX);
                if (createSegment) {
                    const collection = this.schemaBySegment.get(createSegment);
                    if (!collection)
                        return UNKNOWN_TOOL;
                    return this.useCases.createDraft.execute({
                        collection,
                        data: stripReservedArgs(args),
                        authorId: auth.userId,
                        ctx: hookCtx,
                    });
                }
                const updateSegment = extractCollectionSegment(name, UPDATE_DRAFT_PREFIX);
                if (updateSegment) {
                    const collection = this.schemaBySegment.get(updateSegment);
                    if (!collection)
                        return UNKNOWN_TOOL;
                    const id = args["id"];
                    const expected = args["expected_version"];
                    if (typeof id !== "string" || typeof expected !== "number")
                        return MISSING_ARG;
                    // Caller may also call get_entry separately; we don't need
                    // the collection on the chokepoint args because UpdateDraft
                    // looks it up from the existing row.
                    return this.useCases.updateDraft.execute({
                        id,
                        expectedVersion: expected,
                        data: stripReservedArgs(args),
                        ctx: hookCtx,
                    });
                }
                return UNKNOWN_TOOL;
            }
        }
    }
}
const UNKNOWN_TOOL = Symbol("unknown-tool");
const MISSING_ARG = Symbol("missing-arg");
/**
 * Strip the `id` + `expected_version` envelope keys before passing
 * the rest to the chokepoint as `data`. Per-collection update tools
 * mix routing keys (id, expected_version) with authoring fields at
 * the same level; this re-separates them.
 */
const RESERVED_ARG_KEYS = ["id", "expected_version"];
function stripReservedArgs(args) {
    const out = {};
    for (const [k, v] of Object.entries(args)) {
        if (RESERVED_ARG_KEYS.includes(k))
            continue;
        out[k] = v;
    }
    return out;
}
const VIEW_RESERVED_ARG_KEYS = ["page", "show"];
function stripViewReservedArgs(args) {
    const out = {};
    for (const [k, v] of Object.entries(args)) {
        if (VIEW_RESERVED_ARG_KEYS.includes(k))
            continue;
        out[k] = v;
    }
    return out;
}
