// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/infrastructure/mcp/McpResponses.js
import { httpStatusFor, redactForWire, runtimeDiagnostic, } from "@aotterclam/clam-cms-spec";
export function jsonRpcOk(id, result) {
    return new Response(JSON.stringify({ jsonrpc: "2.0", id, result }), {
        headers: { "content-type": "application/json" },
    });
}
export function jsonRpcOkRaw(id, resultJson) {
    const idJson = JSON.stringify(id);
    return new Response(`{"jsonrpc":"2.0","id":${idJson},"result":${resultJson}}`, {
        headers: { "content-type": "application/json" },
    });
}
export function jsonRpcError(id, code, message, data) {
    const error = { code, message };
    if (data !== undefined)
        error.data = data;
    return new Response(JSON.stringify({ jsonrpc: "2.0", id, error }), {
        headers: { "content-type": "application/json" },
    });
}
/**
 * Build a JSON-RPC error `Response` carrying a structured Diagnostic.
 * Used by transport adapters at the layer outside the dispatcher for
 * HTTP-level rejections (missing bearer token, non-staff caller) that
 * never enter the JSON-RPC dispatcher's own catch path.
 */
export function mcpJsonRpcError(diagnostic, requestId = null) {
    const redacted = redactForWire(diagnostic);
    const body = JSON.stringify({
        jsonrpc: "2.0",
        id: requestId,
        error: { code: -32000, message: redacted.message, data: redacted },
    });
    return new Response(body, {
        status: httpStatusFor(diagnostic),
        headers: { "content-type": "application/json" },
    });
}
/** Helper to construct an "unauthenticated" MCP error response. */
export function mcpUnauthenticated(message) {
    return mcpJsonRpcError(runtimeDiagnostic({
        code: "UNAUTHENTICATED",
        severity: "error",
        path: "mcp:fetch",
        expected: "valid bearer token",
        message,
    }));
}
