// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/infrastructure/mcp/McpToolCatalog.js
import { CLAM_BIND_KEYWORD, } from "@aotterclam/clam-cms-spec";
import { mcpToolNameSegment } from "../../domain/service/McpToolNaming.js";
export const MEDIA_TOOLS = [
    {
        name: "create_media_upload",
        description: "Issue a short-lived direct-upload capability for a media object. The caller PUTs the bytes to the returned uploadUrl using the requiredHeaders, then calls commit_media_upload with the same uploadId. Only registered when the runtime has a media storage adapter bound.",
        inputSchema: {
            type: "object",
            properties: {
                filename: { type: "string", description: "Original filename — used in object metadata only; the storage key is server-generated." },
                mimeType: { type: "string", description: "Content-Type. Allowlist: image/png, image/jpeg, image/webp, image/gif. SVG only with adapter opt-in." },
                byteSize: { type: "number", description: "Optional. When provided, enforced in the signed PUT and at commit." },
                alt: { type: "string" },
                caption: { type: "string" },
                purpose: { type: "string", description: "Optional purpose tag (e.g. 'post-cover')." },
            },
            required: ["filename", "mimeType"],
        },
    },
    {
        name: "commit_media_upload",
        description: "Commit a previously-PUT object. Verifies the bytes landed at the storage backend and writes commit metadata. Returns the committed MediaAsset including its publicUrl. Only registered when the runtime has a media storage adapter bound.",
        inputSchema: {
            type: "object",
            properties: {
                uploadId: { type: "string", description: "Returned by create_media_upload." },
                alt: { type: "string" },
                caption: { type: "string" },
                checksum: { type: "string", description: "Optional client-side sha256; verified against storage etag when supplied." },
            },
            required: ["uploadId"],
        },
    },
];
export const GENERIC_TOOLS = [
    {
        name: "list_entries",
        description: "List entries in a collection. Optional filter by status.",
        inputSchema: {
            type: "object",
            properties: {
                collection: { type: "string" },
                status: { type: "string", enum: ["draft", "published", "archived"] },
                limit: { type: "number" },
            },
            required: ["collection"],
        },
    },
    {
        name: "get_entry",
        description: "Fetch a single entry by id.",
        inputSchema: {
            type: "object",
            properties: { id: { type: "string" } },
            required: ["id"],
        },
    },
    {
        name: "request_publish",
        description: "Publish a draft. v0.1.0 simple lifecycle: publishes immediately.",
        inputSchema: {
            type: "object",
            properties: { id: { type: "string" } },
            required: ["id"],
        },
    },
    {
        name: "unpublish_entry",
        description: "Unpublish a published or archived entry back to draft before editing.",
        inputSchema: {
            type: "object",
            properties: { id: { type: "string" } },
            required: ["id"],
        },
    },
    {
        name: "archive_entry",
        description: "Archive an entry. Requires expected_version (OCC).",
        inputSchema: {
            type: "object",
            properties: {
                id: { type: "string" },
                expected_version: { type: "number" },
            },
            required: ["id", "expected_version"],
        },
    },
];
export const CREATE_DRAFT_PREFIX = "create_draft_";
export const UPDATE_DRAFT_PREFIX = "update_draft_";
export const QUERY_VIEW_PREFIX = "query_view_";
export function buildMcpToolCatalog(schemas, opts = {}) {
    const surface = opts.surface ?? "staff";
    if (surface === "public") {
        return (opts.views ?? []).map(buildQueryViewTool);
    }
    const out = [...GENERIC_TOOLS];
    if (opts.mediaEnabled)
        out.push(...MEDIA_TOOLS);
    for (const s of schemas) {
        out.push(buildCreateTool(s));
        out.push(buildUpdateTool(s));
    }
    return out;
}
/** Re-export the naming util from `domain/service/` so existing
 *  consumers of `McpToolCatalog` (the dispatcher) keep their import
 *  surface stable. */
export { mcpToolNameSegment as toolNameSegment };
/** Inverse routing: given a tool name and its prefix, recover the
 *  segment. Returns `null` if the name doesn't carry the prefix. */
export function extractCollectionSegment(toolName, prefix) {
    if (!toolName.startsWith(prefix))
        return null;
    const segment = toolName.slice(prefix.length);
    return segment.length === 0 ? null : segment;
}
function buildCreateTool(schema) {
    const { properties, required } = filterAuthoringFields(schema);
    const inputSchema = {
        type: "object",
        properties,
    };
    if (required.length > 0)
        inputSchema["required"] = required;
    return {
        name: `${CREATE_DRAFT_PREFIX}${mcpToolNameSegment(schema.metadata.name)}`,
        description: describeCreateTool(schema),
        inputSchema,
    };
}
function buildUpdateTool(schema) {
    const { properties, required } = filterAuthoringFields(schema);
    const inputSchema = {
        type: "object",
        properties: {
            id: { type: "string", description: "Entry id to update." },
            expected_version: {
                type: "number",
                description: "OCC version (must match current row version).",
            },
            ...properties,
        },
        required: ["id", "expected_version", ...required],
    };
    return {
        name: `${UPDATE_DRAFT_PREFIX}${mcpToolNameSegment(schema.metadata.name)}`,
        description: describeUpdateTool(schema),
        inputSchema,
    };
}
function buildQueryViewTool(view) {
    const params = view.spec.params;
    const properties = {
        ...(params?.properties ?? {}),
        page: { type: "number", description: "Optional 1-based page number." },
        show: { type: "number", description: "Optional page size, capped by the View limit." },
    };
    const inputSchema = {
        type: "object",
        properties,
    };
    if (params?.required?.length)
        inputSchema["required"] = params.required;
    return {
        name: `${QUERY_VIEW_PREFIX}${mcpToolNameSegment(view.metadata.name)}`,
        description: `Query public View '${view.metadata.name}'.`,
        inputSchema,
    };
}
function describeCreateTool(schema) {
    const base = `Create a new draft entry in '${schema.metadata.name}'.`;
    return schema.spec.description ? `${base} ${schema.spec.description}`.trim() : base;
}
function describeUpdateTool(schema) {
    return `Update a draft entry in '${schema.metadata.name}' with optimistic-concurrency check.`;
}
/**
 * Strip server-stamped properties (those carrying `x-clam-bind`) from
 * the Schema's authoring surface. The agent must not send these — the
 * builtin op projection drops them anyway, but exposing them in the
 * tool schema would invite confusion.
 *
 * The remaining `required` is the intersection of the original
 * `required` array and the surviving property keys.
 */
function filterAuthoringFields(schema) {
    const props = schema.spec.schema.properties ?? {};
    const originalRequired = schema.spec.schema.required ?? [];
    const properties = {};
    const required = [];
    for (const [key, propDef] of Object.entries(props)) {
        if (hasClamBind(propDef))
            continue;
        properties[key] = propDef;
        if (originalRequired.includes(key))
            required.push(key);
    }
    return { properties, required };
}
function hasClamBind(propDef) {
    if (typeof propDef !== "object" || propDef === null)
        return false;
    return CLAM_BIND_KEYWORD in propDef;
}
