// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/procedure/InvokeBuiltinUseCase.js
import { DiagnosticError, runtimeDiagnostic, } from "@aotterclam/clam-cms-spec";
import { projectAndStamp } from "../../domain/service/BuiltinProjector.js";
import { assertEntryWritable } from "../content/EntryWriteGuards.js";
/**
 * `InvokeBuiltinUseCase` — executes the `handler.kind: builtin` op
 * (POC ADR-0014). The four ops map 1:1 to the entry-writer chokepoint:
 *
 *   - `create` → `entries.create({ ..., status: 'draft' })` with a
 *     generated id. Input is projected through
 *     `domain/service/BuiltinProjector.projectAndStamp` so only
 *     Schema-declared keys land in `data` and `x-clam-bind` fields are
 *     server-stamped from `ctx`.
 *   - `update` → `entries.update`. Caller supplies `id` +
 *     `expectedVersion` in the input; OCC enforced at the chokepoint.
 *   - `upsert` → `update` if `input.id` resolves, else `create`.
 *   - `delete` → `entries.delete({ id })`.
 *
 * Pre-projection original input is forwarded to the chokepoint via
 * `originalInput`, so lifecycle hook handlers can read side-channel
 * fields (CAPTCHA tokens, etc.) declared on the Procedure input but
 * not on the Schema.
 *
 * Auth + input/output validation happen upstream in
 * `InvokeProcedureUseCase`. This use case trusts its `validatedInput`.
 */
export class InvokeBuiltinUseCase {
    entries;
    schemasByName;
    clock;
    idgen;
    siteConfig;
    constructor(entries, schemasByName, clock, idgen, siteConfig) {
        this.entries = entries;
        this.schemasByName = schemasByName;
        this.clock = clock;
        this.idgen = idgen;
        this.siteConfig = siteConfig;
    }
    async run(request) {
        const handler = request.procedure.spec.handler;
        if (handler.kind !== "builtin") {
            throw new DiagnosticError(runtimeDiagnostic({
                code: "INTERNAL_ERROR",
                severity: "error",
                path: `usecase/InvokeBuiltin/${request.procedure.metadata.name}`,
                value: handler.kind,
                expected: "handler.kind: 'builtin' (router should have filtered ref)",
                message: `InvokeBuiltinUseCase received handler.kind '${handler.kind}'. The router in InvokeProcedureUseCase should have filtered ref handlers — this is a wiring bug.`,
            }));
        }
        const schema = this.schemasByName.get(handler.schema);
        if (!schema) {
            throw new DiagnosticError(runtimeDiagnostic({
                code: "BUILTIN_HANDLER_SCHEMA_UNKNOWN",
                severity: "error",
                path: `usecase/InvokeBuiltin/${request.procedure.metadata.name}#/handler/schema`,
                value: handler.schema,
                expected: "name of a declared Schema",
                candidates: [...this.schemasByName.keys()],
                message: `Procedure '${request.procedure.metadata.name}' (handler.kind: builtin) targets unknown Schema '${handler.schema}'. (Boot validator should have caught this.)`,
            }));
        }
        const now = this.clock.now();
        const input = request.validatedInput;
        switch (handler.op) {
            case "create":
                return this.opCreate(schema, input, request.ctx, now);
            case "update":
                return this.opUpdate(schema, input, request.ctx, now);
            case "upsert":
                return this.opUpsert(schema, input, request.ctx, now);
            case "delete":
                return this.opDelete(schema, input, request.ctx);
            case "archive":
                return this.opArchive(schema, input, request.ctx, now);
            default: {
                const _exhaustive = handler.op;
                throw new DiagnosticError(runtimeDiagnostic({
                    code: "INTERNAL_ERROR",
                    severity: "error",
                    path: `usecase/InvokeBuiltin/${request.procedure.metadata.name}#/handler/op`,
                    value: _exhaustive,
                    expected: "one of BUILTIN_OPS",
                    message: `Builtin op '${String(_exhaustive)}' is in BUILTIN_OPS but not handled by InvokeBuiltinUseCase. Add a case branch.`,
                }));
            }
        }
    }
    async opCreate(schema, input, ctx, now) {
        const opPath = `usecase/InvokeBuiltin/${schema.metadata.name}/create`;
        const data = projectAndStamp({ schema, input, ctx, clockNow: now });
        await assertEntryWritable({
            opPath,
            entries: this.entries,
            schema,
            data,
            siteConfig: this.siteConfig,
        });
        return this.entries.create({
            id: this.idgen.next(),
            collection: schema.metadata.name,
            status: "draft",
            data,
            authorId: ctx.user?.id ?? null,
            now,
            hookContext: ctx,
            originalInput: input,
        });
    }
    async opUpdate(schema, input, ctx, now) {
        const opPath = `usecase/InvokeBuiltin/${schema.metadata.name}/update`;
        const id = requireField(input, "id", "string");
        const expectedVersion = requireField(input, "expectedVersion", "number");
        const data = projectAndStamp({ schema, input, ctx, clockNow: now });
        await assertEntryWritable({
            opPath,
            entries: this.entries,
            schema,
            data,
            excludeId: id,
            siteConfig: this.siteConfig,
        });
        return this.entries.update({
            id,
            collection: schema.metadata.name,
            expectedVersion,
            data,
            now,
            hookContext: ctx,
            originalInput: input,
        });
    }
    async opUpsert(schema, input, ctx, now) {
        const id = typeof input["id"] === "string" ? input["id"] : undefined;
        if (id) {
            const existing = await this.entries.get(id);
            if (existing)
                return this.opUpdate(schema, input, ctx, now);
        }
        return this.opCreate(schema, input, ctx, now);
    }
    async opDelete(schema, input, ctx) {
        const id = requireField(input, "id", "string");
        return this.entries.delete({
            id,
            collection: schema.metadata.name,
            hookContext: ctx,
            originalInput: input,
        });
    }
    async opArchive(schema, input, ctx, now) {
        const id = requireField(input, "id", "string");
        const expectedVersion = requireField(input, "expectedVersion", "number");
        return this.entries.archive({
            id,
            collection: schema.metadata.name,
            expectedVersion,
            now,
            hookContext: ctx,
            originalInput: input,
        });
    }
}
function requireField(input, key, expected) {
    const v = input[key];
    const ok = expected === "string" ? typeof v === "string" : typeof v === "number" && Number.isFinite(v);
    if (!ok) {
        throw new DiagnosticError(runtimeDiagnostic({
            code: "INPUT_VALIDATION_FAILED",
            severity: "error",
            path: `builtin-input/${key}`,
            value: v,
            expected: `${expected} field '${key}' (declare it in Procedure.input.properties so zod validation rejects upstream)`,
            message: `Builtin op requires '${key}' as a ${expected}. Add it to the Procedure's input schema.`,
        }));
    }
    return v;
}
