// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/procedure/InvokeProcedureUseCase.js
import { DiagnosticError, firstZodIssueAsJsonPointer, jsonSchemaToZod, makeDiagnostic, readJsonPointer, } from "@aotterclam/clam-cms-spec";
/**
 * `InvokeProcedureUseCase` — in-process invocation of a Procedure.
 * HTTP-agnostic — the adapter's mount layer wraps this to map to/from
 * `Request` / `Response`. The test harness exercises the same entry
 * point.
 *
 * Order of operations:
 *   1. Evaluate `requires.auth.all` predicates against `ctx`. Deny ⇒
 *      `AUTH_DENIED`.
 *   2. Validate `input` against the Procedure's `input` JSON Schema.
 *      Fail ⇒ `INPUT_VALIDATION_FAILED` (first zod issue).
 *   3. Dispatch by `handler.kind`:
 *      - `ref`: resolve in registry, call (`InvokeFailure` → unwrap;
 *        other throws → `INTERNAL_ERROR`).
 *      - `builtin`: delegate to `InvokeBuiltinUseCase` (project +
 *        stamp + chokepoint write per POC ADR-0014). When no builtin
 *        collaborator was injected (test paths, ref-only runtimes) the
 *        use case returns `HANDLER_BUILTIN_NOT_IN_V010`.
 *   4. Validate the result against the `output` schema. Fail ⇒
 *      `OUTPUT_VALIDATION_FAILED` (handler / builtin bug).
 */
/**
 * Throwable carrier for handlers that want to surface a structured
 * Diagnostic without going through the generic `INTERNAL_ERROR`
 * envelope. The use case catches it and converts to an
 * `{ ok: false, diagnostic }` return. Lives here (not in
 * `usecase/dto/`) because it's a runtime value, not a DTO type.
 */
export class InvokeFailure extends Error {
    diagnostic;
    constructor(diagnostic) {
        super(diagnostic.message);
        this.diagnostic = diagnostic;
        this.name = "InvokeFailure";
    }
}
export class InvokeProcedureUseCase {
    registry;
    builtin;
    // Per-Procedure compiled zod schemas. zod composition is pure
    // object-tree assembly (no `new Function`) — Workers-CSP-safe.
    // Cached per use-case instance because `jsonSchemaToZod` walks the
    // JSON Schema once per call. Two runtimes with manifests sharing a
    // procedure name + different schemas will get separate caches
    // because the runtime owns the use case instance.
    inputCache = new Map();
    outputCache = new Map();
    constructor(registry, builtin) {
        this.registry = registry;
        this.builtin = builtin;
    }
    async execute(request) {
        const { procedure, input, ctx } = request;
        const phase = request.phase ?? "runtime";
        const procPath = request.pathPrefix ?? `manifest:Procedure/${procedure.metadata.name}`;
        // 1. Auth.
        const denial = evaluateAuthAll(procedure, ctx, procPath, phase);
        if (denial)
            return { ok: false, diagnostic: denial };
        // 2. Input validation.
        const inputValidator = this.compileInput(procedure);
        const inputResult = inputValidator.safeParse(input);
        if (!inputResult.success) {
            const { instancePath, message } = firstZodIssueAsJsonPointer(inputResult.error);
            return {
                ok: false,
                diagnostic: makeDiagnostic({
                    code: "INPUT_VALIDATION_FAILED",
                    phase,
                    severity: "error",
                    path: `${procPath}#/input${instancePath}`,
                    value: readJsonPointer(input, instancePath),
                    expected: message,
                }),
            };
        }
        // 3. Dispatch by handler kind.
        const handlerBinding = procedure.spec.handler;
        const handlerLabel = handlerBinding.kind === "builtin"
            ? `builtin/${handlerBinding.op}`
            : handlerBinding.ref;
        let result;
        try {
            if (handlerBinding.kind === "builtin") {
                if (!this.builtin) {
                    return {
                        ok: false,
                        diagnostic: makeDiagnostic({
                            code: "HANDLER_BUILTIN_NOT_IN_V010",
                            phase,
                            severity: "error",
                            path: `${procPath}#/handler/kind`,
                            value: handlerBinding.kind,
                            expected: "InvokeBuiltinUseCase wired into the runtime",
                            message: `Procedure '${procedure.metadata.name}' uses handler.kind: 'builtin' but the runtime was constructed without an InvokeBuiltinUseCase.`,
                        }),
                    };
                }
                result = await this.builtin.run({
                    procedure,
                    validatedInput: inputResult.data,
                    ctx,
                });
            }
            else {
                const handler = this.registry.get(handlerBinding.ref);
                if (!handler) {
                    return {
                        ok: false,
                        diagnostic: makeDiagnostic({
                            code: "HANDLER_NOT_REGISTERED",
                            phase,
                            severity: "error",
                            path: `${procPath}#/handler/ref`,
                            value: handlerBinding.ref,
                            expected: "a ref registered via the handlers option / sdk.registerHandler",
                            candidates: this.registry.list(),
                        }),
                    };
                }
                result = await handler(inputResult.data, ctx);
            }
        }
        catch (err) {
            if (err instanceof InvokeFailure) {
                return { ok: false, diagnostic: err.diagnostic };
            }
            if (err instanceof DiagnosticError) {
                return { ok: false, diagnostic: err.diagnostic };
            }
            const msg = err instanceof Error ? err.message : String(err);
            return {
                ok: false,
                diagnostic: makeDiagnostic({
                    code: "INTERNAL_ERROR",
                    phase,
                    severity: "error",
                    path: procPath,
                    expected: "handler completes without throwing",
                    message: `Handler '${handlerLabel}' threw: ${msg}`,
                }),
            };
        }
        // 4. Output validation.
        const outputValidator = this.compileOutput(procedure);
        const outputResult = outputValidator.safeParse(result);
        if (!outputResult.success) {
            const { instancePath, message } = firstZodIssueAsJsonPointer(outputResult.error);
            return {
                ok: false,
                diagnostic: makeDiagnostic({
                    code: "OUTPUT_VALIDATION_FAILED",
                    phase,
                    severity: "error",
                    path: `${procPath}#/output${instancePath}`,
                    value: readJsonPointer(result, instancePath),
                    expected: message,
                    message: `Handler '${handlerLabel}' returned a value that does not match its declared output schema. This is a handler bug.`,
                }),
            };
        }
        return { ok: true, data: result };
    }
    /** Test seam — clear validator caches between tests when manifests
     *  change between runs. Production paths never call this. */
    _clearValidatorCaches() {
        this.inputCache.clear();
        this.outputCache.clear();
    }
    compileInput(p) {
        const k = p.metadata.name;
        let v = this.inputCache.get(k);
        if (!v) {
            v = jsonSchemaToZod(p.spec.input);
            this.inputCache.set(k, v);
        }
        return v;
    }
    compileOutput(p) {
        const k = p.metadata.name;
        let v = this.outputCache.get(k);
        if (!v) {
            v = jsonSchemaToZod(p.spec.output);
            this.outputCache.set(k, v);
        }
        return v;
    }
}
function evaluateAuthAll(procedure, ctx, procPath, phase) {
    const all = procedure.spec.requires?.auth?.all;
    if (!all || all.length === 0)
        return null;
    for (let i = 0; i < all.length; i++) {
        const pred = all[i];
        if (!evaluatePredicate(pred, ctx)) {
            return makeDiagnostic({
                code: "AUTH_DENIED",
                phase,
                severity: "error",
                path: `${procPath}#/requires/auth/all/${i}`,
                expected: describePredicate(pred),
                message: `Authorization predicate not satisfied: ${describePredicate(pred)}.`,
            });
        }
    }
    return null;
}
function evaluatePredicate(pred, ctx) {
    if (pred === "ctx.user")
        return ctx.user !== null;
    if (typeof pred === "object" && pred !== null && "ctx.staff" in pred) {
        if (!ctx.staff)
            return false;
        return pred["ctx.staff"].includes(ctx.staff.role);
    }
    return false;
}
function describePredicate(pred) {
    if (pred === "ctx.user")
        return "caller is signed in (ctx.user)";
    return `caller is staff with role in [${pred["ctx.staff"].join(", ")}]`;
}
