// @aotterclam/clam-cms-cloudflare@0.0.8-alpha/dist/auth/createAuth.js
import { betterAuth } from "better-auth";
import { admin, mcp } from "better-auth/plugins";
import { createAccessControl } from "better-auth/plugins/access";
import { defaultStatements } from "better-auth/plugins/admin/access";
export const ADMIN_ROLES = ["contributor", "editor", "owner"];
export const ADMIN_ROLE_SET = new Set(ADMIN_ROLES);
const ac = createAccessControl(defaultStatements);
const ownerAc = ac.newRole({
    user: defaultStatements.user,
    session: defaultStatements.session,
});
const editorAc = ac.newRole({
    user: ["list", "ban", "get", "update"],
    session: ["list", "revoke"],
});
const contributorAc = ac.newRole({
    user: ["list", "get"],
    session: [],
});
const userAc = ac.newRole({
    user: [],
    session: [],
});
function buildAuth(config) {
    const socialProviders = {};
    if (config.github) {
        socialProviders.github = {
            clientId: config.github.clientId,
            clientSecret: config.github.clientSecret,
            mapProfileToUser: (profile) => ({
                githubLogin: profile.login,
            }),
            ...(config.github.redirectURI
                ? { redirectURI: config.github.redirectURI }
                : {}),
        };
    }
    const adminGithubLogin = config.adminGithubLogin?.trim();
    return betterAuth({
        database: config.database,
        secret: config.secret,
        baseURL: config.baseURL,
        socialProviders,
        user: {
            additionalFields: {
                githubLogin: {
                    type: "string",
                    required: false,
                    input: false,
                },
            },
        },
        plugins: [
            admin({
                defaultRole: "user",
                adminRoles: [...ADMIN_ROLES],
                ac,
                roles: {
                    owner: ownerAc,
                    editor: editorAc,
                    contributor: contributorAc,
                    user: userAc,
                },
            }),
            mcp({
                loginPage: "/admin/sign-in",
                oidcConfig: {
                    loginPage: "/admin/sign-in",
                    scopes: ["mcp:read", "mcp:staff"],
                    defaultScope: "openid profile email mcp:read",
                    metadata: {
                        scopes_supported: [
                            "openid",
                            "profile",
                            "email",
                            "offline_access",
                            "mcp:read",
                            "mcp:staff",
                        ],
                    },
                },
            }),
        ],
        databaseHooks: {
            user: {
                create: {
                    after: async (user) => {
                        if (!adminGithubLogin)
                            return;
                        const u = user;
                        if (u.githubLogin?.toLowerCase() !== adminGithubLogin.toLowerCase())
                            return;
                        const placeholders = ADMIN_ROLES.map(() => "?").join(",");
                        const existingAdmin = await config.database
                            .prepare(`SELECT id FROM user WHERE role IN (${placeholders}) LIMIT 1`)
                            .bind(...ADMIN_ROLES)
                            .first();
                        if (existingAdmin)
                            return;
                        await config.database
                            .prepare("UPDATE user SET role = ? WHERE id = ?")
                            .bind("owner", u.id)
                            .run();
                    },
                },
            },
        },
    });
}
export function createAuth(config) {
    const auth = buildAuth(config);
    // The mcp plugin is always loaded by buildAuth, so getMcpSession is
    // guaranteed present. Bind once at construction.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const api = auth.api;
    return {
        handler: (request) => auth.handler(request),
        getSession: (request) => api.getSession({ headers: request.headers }).then((r) => r ?? null),
        getMcpSession: async (request) => {
            const r = await api.getMcpSession({ headers: request.headers });
            if (!r)
                return null;
            const raw = r;
            return {
                ...raw,
                scopes: Array.isArray(raw.scopes)
                    ? raw.scopes
                    : raw.scopes.split(/\s+/).filter(Boolean),
            };
        },
        getUserRole: async (userId) => {
            const row = await config.database
                .prepare("SELECT role FROM user WHERE id = ? LIMIT 1")
                .bind(userId)
                .first();
            return row?.role ?? null;
        },
    };
}
