// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/index.js
/**
 * `@aotterclam/clam-cms-runtime` — public surface.
 *
 * Adapter-agnostic runtime engine for clam-cms. Layered per the
 * Aotter clean-architecture convention (mirrors `aotter-clam/clam/core`):
 *
 *   domain ← usecase ← infrastructure ← runtime.ts (assembly root)
 *
 * Adapters (e.g. `@aotterclam/clam-cms-cloudflare`) implement the
 * required port interfaces in `domain/port/` and call `createCmsRuntime`
 * to compose everything. Optional feature ports (for example media
 * hosting) stay adapter-agnostic and are only wired when enabled.
 *
 * MUST NOT import `D1Database` / `KVNamespace` / any Cloudflare-
 * specific type. The Netlify stub package exists as a public reminder.
 */
// Assembly root.
export { createCmsRuntime, } from "./runtime.js";
export { extensionForMime } from "./usecase/media/mediaAllowlist.js";
// ID source — adapters wire this into binding-side helpers that need
// random IDs. Default `RandomUuidGenerator` (`crypto.randomUUID()`)
// can be swapped in tests with a deterministic counter.
export { RandomUuidGenerator } from "./domain/port/IdGenerator.js";
export { TemplateRegistry, } from "./domain/model/TemplateRegistry.js";
export { createPublicPathResolver, } from "./domain/service/PublicPathResolver.js";
export { composeEntrySeoMeta, renderSeoTagsHtml, } from "./domain/service/SeoMetaComposer.js";
// Public route / starter fixture helpers. These are intentionally
// exported one-by-one instead of exposing the whole service barrel.
export { entryHtmlKey, entryMarkdownKey, entryHtmlKeyFromParts, entryMarkdownKeyFromParts, listHtmlKey, llmsTxtKey, } from "./domain/service/PublishKeys.js";
export { serializeEntryAsMarkdown } from "./domain/service/MarkdownSerializer.js";
export { readEntryBySlug } from "./domain/service/PublishedEntries.js";
export { inferLocaleFromPath, isKnownLocale, siteConfigFromDefaults, toUrlLocale, } from "./domain/service/LocaleNegotiator.js";
export { matchPath } from "./domain/service/PathMatcher.js";
export { coerceViewParams, ViewParamCoercionError, } from "./domain/service/ViewParamCoercer.js";
// Explicit adapter contract for the Cloudflare MCP mount. Do not
// export the whole MCP infrastructure barrel from the root.
export { McpJsonRpcDispatcher, } from "./infrastructure/mcp/McpJsonRpcDispatcher.js";
// Starter fixture support. Kept explicit so persistence/http
// infrastructure do not become root public API by accident.
export { CANONICAL_MIGRATIONS } from "./infrastructure/boot/canonicalMigrations.js";
// Procedure handler failure carrier used by platform helper handlers
// such as Cloudflare Turnstile.
export { InvokeFailure } from "./usecase/procedure/InvokeProcedureUseCase.js";
