// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/runtime.js
import { partitionManifests, } from "@aotterclam/clam-cms-spec";
import { SystemClock } from "./domain/port/Clock.js";
import { RandomUuidGenerator, } from "./domain/port/IdGenerator.js";
import { buildHandlerRegistry, } from "./domain/port/HandlerRegistry.js";
import { ArchiveUseCase, CreateDraftUseCase, DeleteEntryUseCase, GetEntryUseCase, ListEntriesUseCase, RequestPublishUseCase, UnpublishUseCase, UpdateDraftUseCase, } from "./usecase/content/index.js";
import { InvokeBuiltinUseCase, InvokeProcedureUseCase, } from "./usecase/procedure/index.js";
import { ExecuteViewUseCase } from "./usecase/view/index.js";
import { ValidateBootUseCase } from "./usecase/boot/index.js";
import { RunDeferredHookUseCase, RunLifecycleHooksUseCase, } from "./usecase/lifecycle/index.js";
import { ComposeEntrySeoMetaUseCase, ComposeLlmsTxtUseCase, ComposeSitemapUseCase, PreviewEntryUseCase, RenderEntryLiveUseCase, RenderListLiveUseCase, } from "./usecase/render/index.js";
import { CommitMediaUploadUseCase, CreateMediaUploadUseCase, } from "./usecase/media/index.js";
import { TemplateRegistry as TemplateRegistryImpl } from "./domain/model/TemplateRegistry.js";
import { TriggerIndex } from "./domain/service/TriggerIndex.js";
import { DatabaseEntryRepository } from "./infrastructure/persistence/DatabaseEntryRepository.js";
import { DatabaseSiteConfigRepository } from "./infrastructure/persistence/DatabaseSiteConfigRepository.js";
import { LifecycleHookingEntryRepository } from "./infrastructure/persistence/LifecycleHookingEntryRepository.js";
import { HtmlPublishOrchestrator } from "./infrastructure/render/index.js";
import { CANONICAL_MIGRATIONS } from "./infrastructure/boot/index.js";
export function createCmsRuntime(args) {
    const partitioned = partitionManifests([...args.manifests]);
    const schemasByName = new Map();
    for (const s of partitioned.schemas)
        schemasByName.set(s.metadata.name, s);
    const proceduresByName = new Map();
    for (const p of partitioned.procedures)
        proceduresByName.set(p.metadata.name, p);
    const viewsByName = new Map();
    for (const v of partitioned.views)
        viewsByName.set(v.metadata.name, v);
    const triggersByName = new Map();
    for (const t of partitioned.triggers)
        triggersByName.set(t.metadata.name, t);
    const registry = buildHandlerRegistry(args.handlers ?? {});
    const templates = args.templates ?? new TemplateRegistryImpl();
    const clock = args.clock ?? SystemClock;
    const idgen = args.idgen ?? RandomUuidGenerator;
    // Repositories: DB-backed inner + lifecycle-hook decorator. Every
    // mutation through `entries` (create / update / delete / archive /
    // transitionStatus) fires the matching Triggers via
    // `RunLifecycleHooksUseCase`. Symmetric chokepoint per POC ADR-0014:
    // MCP, admin, and builtin paths all hit the same wrapped repository.
    const innerEntries = new DatabaseEntryRepository(args.db);
    const triggerIndex = new TriggerIndex(partitioned.triggers);
    const siteConfig = new DatabaseSiteConfigRepository(args.db);
    // `entries` is filled below — assigned via `let` so the lifecycle
    // hooks (which run procedures, which can themselves write entries
    // via builtin handlers) close over the wrapped repo, not the bare
    // DB-backed one. Without this every builtin write inside a hook
    // would skip the decorator and silently bypass downstream hooks.
    let entries;
    const entriesProxy = {
        create: (a) => entries.create(a),
        get: (id) => entries.get(id),
        update: (a) => entries.update(a),
        delete: (a) => entries.delete(a),
        archive: (a) => entries.archive(a),
        transitionStatus: (a) => entries.transitionStatus(a),
        list: (a) => entries.list(a),
        findByDataField: (a) => entries.findByDataField(a),
        findByDataFields: (a) => entries.findByDataFields(a),
    };
    const invokeBuiltin = new InvokeBuiltinUseCase(entriesProxy, schemasByName, clock, idgen, siteConfig);
    const invokeProcedure = new InvokeProcedureUseCase(registry, invokeBuiltin);
    const lifecycleHooks = new RunLifecycleHooksUseCase(triggerIndex, proceduresByName, invokeProcedure);
    entries = new LifecycleHookingEntryRepository(innerEntries, triggerIndex, lifecycleHooks, args.deferredHookDispatcher);
    const runDeferredHook = new RunDeferredHookUseCase(lifecycleHooks);
    const publicPathResolver = args.publicPathResolver ?? null;
    const composeEntrySeoMeta = new ComposeEntrySeoMetaUseCase(args.db);
    const publishOrchestrator = new HtmlPublishOrchestrator(args.db, args.kv, publicPathResolver, composeEntrySeoMeta);
    // Content / view / boot use cases. They see `entries` only as the
    // chokepoint port — hook firing is invisible to them.
    const createDraft = new CreateDraftUseCase(entries, schemasByName, clock, idgen, siteConfig);
    const updateDraft = new UpdateDraftUseCase(entries, schemasByName, clock, siteConfig);
    const getEntry = new GetEntryUseCase(entries);
    const listEntries = new ListEntriesUseCase(entries, schemasByName);
    const contentPublishEffects = { publishOrchestrator, siteConfig, templates };
    const requestPublish = new RequestPublishUseCase(entries, schemasByName, clock, contentPublishEffects, siteConfig);
    const unpublish = new UnpublishUseCase(entries, clock, contentPublishEffects);
    const archive = new ArchiveUseCase(entries, schemasByName, clock, contentPublishEffects);
    const deleteEntry = new DeleteEntryUseCase(entries);
    const executeView = new ExecuteViewUseCase(args.db);
    const composeLlmsTxt = new ComposeLlmsTxtUseCase(args.db);
    const composeSitemap = new ComposeSitemapUseCase(args.db);
    const renderEntryLive = new RenderEntryLiveUseCase(args.db, templates, publicPathResolver, composeEntrySeoMeta);
    const renderListLive = new RenderListLiveUseCase(args.db, templates);
    const previewEntry = new PreviewEntryUseCase(args.db, templates, publicPathResolver, composeEntrySeoMeta);
    const validateBoot = new ValidateBootUseCase();
    const media = args.mediaStorage
        ? {
            storage: args.mediaStorage,
            createUpload: new CreateMediaUploadUseCase(args.mediaStorage, args.kv, clock, { allowSvg: args.mediaAllowSvg ?? false, maxBytes: args.mediaMaxBytes }),
            commitUpload: new CommitMediaUploadUseCase(args.mediaStorage, args.kv, clock, { maxBytes: args.mediaMaxBytes }),
        }
        : null;
    return {
        db: args.db,
        kv: args.kv,
        assets: args.assets,
        createDraft,
        updateDraft,
        getEntry,
        listEntries,
        requestPublish,
        unpublish,
        archive,
        deleteEntry,
        invokeProcedure,
        executeView,
        composeLlmsTxt,
        composeSitemap,
        composeEntrySeoMeta,
        renderEntryLive,
        renderListLive,
        previewEntry,
        validateBoot,
        publishOrchestrator,
        siteConfig,
        publicPathResolver,
        media,
        runDeferredHook,
        registry,
        templates,
        schemasByName,
        proceduresByName,
        viewsByName,
        triggers: partitioned.triggers,
        triggersByName,
        clock,
        idgen,
        async bootInit() {
            await args.db.migrations.runAll(CANONICAL_MIGRATIONS);
            await siteConfig.seed(args.siteDefaults);
            const siteLocales = await siteConfig.readLocales();
            validateBoot.assert({
                manifests: args.manifests,
                registry,
                siteLocales,
            });
        },
    };
}
