// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/boot/ValidateBootUseCase.js
import { bootDiagnostic, partitionManifests, checkLocaleAndTranslates, } from "@aotterclam/clam-cms-spec";
import { mcpToolNameSegment } from "../../domain/service/McpToolNaming.js";
export class ValidateBootUseCase {
    execute(request) {
        const partitioned = partitionManifests([...request.manifests]);
        const proceduresByName = new Map();
        for (const p of partitioned.procedures)
            proceduresByName.set(p.metadata.name, p);
        const schemasByName = new Map();
        for (const s of partitioned.schemas)
            schemasByName.set(s.metadata.name, s);
        const diagnostics = [];
        const procedureCandidates = [...proceduresByName.keys()];
        const schemaCandidates = [...schemasByName.keys()];
        const handlerCandidates = request.registry.list();
        // 1. Procedure handler refs + builtin schema cross-resolution.
        for (const p of partitioned.procedures) {
            const h = p.spec.handler;
            if (h.kind === "ref" && !request.registry.has(h.ref)) {
                diagnostics.push(bootDiagnostic({
                    code: "HANDLER_NOT_REGISTERED",
                    severity: "error",
                    path: `manifest:Procedure/${p.metadata.name}#/spec/handler/ref`,
                    value: h.ref,
                    expected: `a function registered via the handlers option / sdk.registerHandler('${h.ref}', fn)`,
                    candidates: handlerCandidates,
                    message: `Procedure '${p.metadata.name}' declares handler.ref '${h.ref}' but no handler is registered for that key. Wire it in your project's handlers map: { '${h.ref}': ... }.`,
                }));
            }
            if (h.kind === "builtin" && !schemasByName.has(h.schema)) {
                diagnostics.push(bootDiagnostic({
                    code: "BUILTIN_HANDLER_SCHEMA_UNKNOWN",
                    severity: "error",
                    path: `manifest:Procedure/${p.metadata.name}#/spec/handler/schema`,
                    value: h.schema,
                    expected: "name of a declared Schema",
                    candidates: schemaCandidates,
                    message: `Procedure '${p.metadata.name}' (handler.kind: builtin) targets unknown Schema '${h.schema}'.`,
                }));
            }
        }
        // 2. Trigger.target.procedure resolves.
        for (const t of partitioned.triggers) {
            if (!proceduresByName.has(t.spec.target.procedure)) {
                diagnostics.push(bootDiagnostic({
                    code: "TRIGGER_TARGET_PROCEDURE_UNKNOWN",
                    severity: "error",
                    path: `manifest:Trigger/${t.metadata.name}#/spec/target/procedure`,
                    value: t.spec.target.procedure,
                    expected: "name of a declared Procedure",
                    candidates: procedureCandidates,
                    message: `Trigger '${t.metadata.name}' targets unknown Procedure '${t.spec.target.procedure}'.`,
                }));
            }
            if (t.spec.source.kind === "lifecycle") {
                if (!schemasByName.has(t.spec.source.schema)) {
                    diagnostics.push(bootDiagnostic({
                        code: "LIFECYCLE_SCHEMA_UNKNOWN",
                        severity: "error",
                        path: `manifest:Trigger/${t.metadata.name}#/spec/source/schema`,
                        value: t.spec.source.schema,
                        expected: "name of a declared Schema",
                        candidates: schemaCandidates,
                        message: `Trigger '${t.metadata.name}' watches unknown Schema '${t.spec.source.schema}'.`,
                    }));
                }
            }
        }
        // 3. HTTP trigger uniqueness + /api/ prefix.
        diagnostics.push(...checkHttpRouteCollisions(partitioned.triggers));
        diagnostics.push(...checkHttpRoutePrefix(partitioned.triggers));
        // 4. MCP tool-name collision (POC PR #48): per-collection tool
        //    emission lowercases + kebab→snake the Schema name; two
        //    Schemas that mangle to the same suffix would silently
        //    overwrite each other in `tools/list`.
        diagnostics.push(...checkMcpToolNameCollisions(partitioned.schemas));
        // 5. Locale + translates cross-Schema invariants.
        diagnostics.push(...checkLocaleAndTranslates({
            schemas: partitioned.schemas,
            phase: "boot",
            siteLocales: request.siteLocales,
        }));
        if (diagnostics.length === 0)
            return { ok: true };
        return { ok: false, diagnostics };
    }
    /** Validate; on failure, throw `BootValidationError`. */
    assert(request) {
        const result = this.execute(request);
        if (!result.ok)
            throw new BootValidationError(result.diagnostics);
    }
}
function checkHttpRouteCollisions(triggers) {
    const seen = new Map();
    const out = [];
    for (const t of triggers) {
        if (t.spec.source.kind !== "http")
            continue;
        const key = `${t.spec.source.method} ${t.spec.source.path}`;
        const prior = seen.get(key);
        if (prior) {
            out.push(bootDiagnostic({
                code: "TRIGGER_PATH_COLLISION",
                severity: "error",
                path: `manifest:Trigger/${t.metadata.name}#/spec/source`,
                value: key,
                expected: `unique (method, path) across http Triggers (also declared by '${prior}')`,
                message: `Trigger '${t.metadata.name}' shares route ${key} with Trigger '${prior}'.`,
            }));
        }
        else {
            seen.set(key, t.metadata.name);
        }
    }
    return out;
}
function checkHttpRoutePrefix(triggers) {
    const out = [];
    for (const t of triggers) {
        if (t.spec.source.kind !== "http")
            continue;
        const path = t.spec.source.path;
        if (!path.startsWith("/api/")) {
            out.push(bootDiagnostic({
                code: "TRIGGER_PATH_INVALID",
                severity: "error",
                path: `manifest:Trigger/${t.metadata.name}#/spec/source/path`,
                value: path,
                expected: "path starting with '/api/'",
                message: `Trigger '${t.metadata.name}' has path '${path}' — http Trigger ` +
                    `paths MUST start with '/api/' so adapters can route public ` +
                    `pages and Procedure endpoints without ambiguity.`,
            }));
        }
    }
    return out;
}
function checkMcpToolNameCollisions(schemas) {
    const seen = new Map();
    const out = [];
    for (const s of schemas) {
        const segment = mcpToolNameSegment(s.metadata.name);
        const prior = seen.get(segment);
        if (prior && prior !== s.metadata.name) {
            out.push(bootDiagnostic({
                code: "MCP_TOOL_NAME_COLLISION",
                severity: "error",
                path: `manifest:Schema/${s.metadata.name}#/metadata/name`,
                value: segment,
                expected: `Schema name unique after kebab→snake mangling (collides with '${prior}')`,
                message: `Schema '${s.metadata.name}' mangles to MCP tool suffix '${segment}', ` +
                    `which already comes from Schema '${prior}'. Rename one of the Schemas ` +
                    `(e.g. avoid mixing '${prior}' and '${s.metadata.name}') so per-collection ` +
                    `MCP tool emission stays unambiguous.`,
            }));
        }
        else if (!prior) {
            seen.set(segment, s.metadata.name);
        }
    }
    return out;
}
/**
 * Single Error wrapping every boot diagnostic. The runtime throws this
 * during `bootInit`; adapters surface it in their init logs.
 */
export class BootValidationError extends Error {
    diagnostics;
    constructor(diagnostics) {
        const summary = diagnostics
            .map((d) => `  - ${d.code} (phase: ${d.phase}) at ${d.path}: ${d.message}`)
            .join("\n");
        super(`Runtime boot validation failed (${diagnostics.length} error(s)):\n${summary}\n\n` +
            `See ADR-0007 (boot-time fail-fast) for context. Diagnostics also ` +
            `available on the .diagnostics field of this error for programmatic handling.`);
        this.diagnostics = diagnostics;
        this.name = "BootValidationError";
    }
}
