// @aotterclam/clam-cms-cloudflare@0.0.8-alpha/dist/bindings/D1DatabaseDriver.js
/**
 * `DatabaseDriver` impl wrapping Cloudflare's `D1Database` binding.
 *
 * The runtime port shape is intentionally close to D1's API (which is
 * itself close to the SQLite C API), so this wrapper is mostly a
 * type-narrowing pass-through. The notable behaviour is the
 * `D1Migrations` runner: it records applied migrations in a
 * `_clam_migrations` table so subsequent boots are idempotent.
 *
 * Per ADR-0011 the runtime never imports `D1Database` itself — this
 * file is the only place in the codebase that does.
 */
export class D1DatabaseDriver {
    db;
    migrations;
    constructor(db) {
        this.db = db;
        this.migrations = new D1Migrations(db);
    }
    prepare(sql) {
        return wrap(this.db.prepare(sql));
    }
    async batch(stmts) {
        const native = stmts.map((s) => unwrap(s));
        const results = await this.db.batch(native);
        return results.map((r) => ({
            success: r.success,
            meta: { changes: r.meta.changes ?? 0 },
            results: r.results,
        }));
    }
}
const NATIVE = new WeakMap();
function wrap(native) {
    const stmt = {
        bind: (...params) => wrap(native.bind(...params)),
        first: () => native.first().then((v) => v ?? null),
        all: async () => {
            const r = await native.all();
            return (r.results ?? []);
        },
        run: async () => {
            const r = await native.run();
            return {
                success: r.success,
                meta: { changes: r.meta.changes ?? 0, lastRowId: r.meta.last_row_id },
            };
        },
    };
    NATIVE.set(stmt, native);
    return stmt;
}
function unwrap(stmt) {
    const native = NATIVE.get(stmt);
    if (!native) {
        throw new Error("D1DatabaseDriver.batch received a statement not produced by this driver. " +
            "Build statements via `db.prepare(...)` on the same DatabaseDriver instance.");
    }
    return native;
}
/**
 * Migration runner. Each migration runs in its own batch — the batch
 * includes the migration's SQL plus the `_clam_migrations` row insert,
 * so a single migration is all-or-nothing. Migrations across boots are
 * NOT transactional: a worker restart between migration N and N+1
 * leaves N applied, N+1 retried on the next boot. Authors must keep
 * each migration's SQL idempotent on retry (use `IF NOT EXISTS`,
 * `INSERT … ON CONFLICT DO NOTHING`, etc.).
 */
class D1Migrations {
    db;
    constructor(db) {
        this.db = db;
    }
    async runAll(migrations) {
        await this.db
            .prepare(`CREATE TABLE IF NOT EXISTS _clam_migrations (id TEXT PRIMARY KEY, applied_at INTEGER NOT NULL)`)
            .run();
        const applied = await this.db
            .prepare(`SELECT id FROM _clam_migrations`)
            .all();
        const seen = new Set((applied.results ?? []).map((r) => r.id));
        for (const m of migrations) {
            if (seen.has(m.id))
                continue;
            const statements = splitSql(m.sql);
            const ops = statements.map((s) => this.db.prepare(s));
            ops.push(this.db
                .prepare(`INSERT INTO _clam_migrations (id, applied_at) VALUES (?, ?)`)
                .bind(m.id, Date.now()));
            await this.db.batch(ops);
        }
    }
}
/**
 * D1 / SQLite accept multi-statement SQL on the wire, but `.batch`
 * expects each statement separately. Splitting on `;` is correct for
 * our migration corpus (DDL + INSERT seeds, no string literals
 * containing `;`). When that ever stops being true, switch to a
 * lexer-aware split.
 */
function splitSql(sql) {
    return sql
        .split(";")
        .map((s) => s.trim())
        .filter((s) => s.length > 0);
}
