// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/usecase/view/ExecuteViewUseCase.js
import { runtimeDiagnostic, } from "@aotterclam/clam-cms-spec";
import { compileView } from "../../domain/service/ViewSqlCompiler.js";
export class ExecuteViewUseCase {
    db;
    constructor(db) {
        this.db = db;
    }
    async execute(request) {
        const viewPath = request.pathPrefix ?? `manifest:View/${request.view.metadata.name}`;
        let compiled;
        try {
            compiled = compileView(request.view, request.options);
        }
        catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            return {
                ok: false,
                diagnostic: runtimeDiagnostic({
                    code: "INTERNAL_ERROR",
                    severity: "error",
                    path: viewPath,
                    expected: "View compiles to valid SQL",
                    message: `View compile failed: ${msg}`,
                }),
            };
        }
        try {
            const rows = await this.db
                .prepare(compiled.sql)
                .bind(...compiled.params)
                .all();
            return {
                ok: true,
                result: {
                    rows,
                    page: compiled.effectivePage,
                    show: compiled.effectiveShow,
                    hasMore: rows.length === compiled.effectiveShow,
                },
            };
        }
        catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            return {
                ok: false,
                diagnostic: runtimeDiagnostic({
                    code: "INTERNAL_ERROR",
                    severity: "error",
                    path: viewPath,
                    expected: "SQL executes without error",
                    message: `View SQL execution failed: ${msg}`,
                }),
            };
        }
    }
}
