// @aotterclam/clam-cms-runtime@0.0.8-alpha/dist/domain/model/EntryRow.js
/**
 * Lift `data.locale` to the top-level `locale?` field. Centralised so
 * every repository impl (DatabaseDriver-backed, in-memory test fake,
 * future adapters) derives the convenience field the same way —
 * adding a per-impl hydration helper would let them drift.
 */
export function liftLocale(data) {
    const v = data["locale"];
    return typeof v === "string" ? v : undefined;
}
export class EntryVersionConflict extends Error {
    id;
    expected;
    actual;
    constructor(id, expected, actual) {
        super(`version conflict on ${id}: expected ${expected}, found ${actual}`);
        this.id = id;
        this.expected = expected;
        this.actual = actual;
        this.name = "EntryVersionConflict";
    }
}
/**
 * Thrown by `transitionStatus({ expectedStatus })` when the row's
 * current status doesn't match. Distinct from version conflict
 * because it carries `ContentState` instead of numeric version, so
 * the orchestration layer can emit a tailored CONFLICT diagnostic
 * (e.g. "expected 'draft', found 'published'").
 */
export class EntryStatusConflict extends Error {
    id;
    expected;
    actual;
    constructor(id, expected, actual) {
        super(`status conflict on ${id}: expected ${expected}, found ${actual}`);
        this.id = id;
        this.expected = expected;
        this.actual = actual;
        this.name = "EntryStatusConflict";
    }
}
